{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "066e0a41-7028-4a7e-ab10-6dce79d013c7",
   "metadata": {},
   "source": [
    "# Subgroup/Class Tree Sequences Preparation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "12c87eb4-36c3-4462-b087-2e6ea0e98dcf",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "from tqdm import tqdm\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "from tqdm import tqdm\n",
    "import os\n",
    "import shutil\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c5a66752-a627-46f0-af87-781b22b0653b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def copy_files(file_list, source_dir, destination_dir):\n",
    "    \"\"\"\n",
    "    Copies files from the source directory to the destination directory based on the provided list of filenames.\n",
    "\n",
    "    :param file_list: List of filenames to copy.\n",
    "    :param source_dir: Directory where the files are currently located.\n",
    "    :param destination_dir: Directory where the files should be copied to.\n",
    "    \"\"\"\n",
    "    # Ensure the destination directory exists\n",
    "    os.makedirs(destination_dir, exist_ok=True)\n",
    "    \n",
    "    # Iterate over the file list with tqdm progress bar\n",
    "    for filename in tqdm(file_list, desc=\"Copying files\"):\n",
    "        # Construct full paths for the source and destination\n",
    "        source_path = os.path.join(source_dir, filename)\n",
    "        destination_path = os.path.join(destination_dir, filename)\n",
    "        \n",
    "        # Copy the file from the source to the destination\n",
    "        shutil.copy(source_path, destination_path)\n",
    "\n",
    "# Example usage:\n",
    "# copy_files(['file1.txt', 'file2.txt'], '/path/to/source', '/path/to/destination')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "bdd551d6-bd71-4d99-81e0-cc2ca373b413",
   "metadata": {},
   "outputs": [],
   "source": [
    "nr_genomes_dataset = pd.read_csv('/Users/akshayonly/Work/04-Complex-I/Data/02-Genomic-Records-and-Metadata/02-Genome-Metadata/genomes_dataset_unique.csv')\n",
    "genomes_dataset = pd.read_csv('/Users/akshayonly/Work/04-Complex-I/Data/02-Genomic-Records-and-Metadata/02-Genome-Metadata/genomes_dataset.csv')\n",
    "taxonomy = pd.read_csv('/Users/akshayonly/Work/04-Complex-I/Data/02-Genomic-Records-and-Metadata/03-Taxonomy-Metadata/taxonomy.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "177496ea-53a9-4bdc-8cbb-c713572e2b0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonomy = genomes_dataset[['Group', 'SubGroup', 'Species']].merge(taxonomy).drop_duplicates()\n",
    "\n",
    "taxonomy = taxonomy[~taxonomy['Class'].isnull()]\n",
    "\n",
    "taxonomy = taxonomy[(~taxonomy['Class'].str.startswith('Candidatus')) & (~taxonomy['Class'].str.startswith('candidatus'))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "434c5d3a-dcb1-48b0-949e-2520b616335c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(65, 100)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxonomy['SubGroup'].nunique(), taxonomy['Class'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "621700dd-7755-46c9-9b57-2d0ddfee30f2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "representative_species = {\n",
    "    \"Acidimicrobiia\": \"Acidimicrobium ferrooxidans\",\n",
    "    \"Acidithiobacillia\": \"Acidithiobacillus ferrooxidans\",\n",
    "    \"Actinomycetes\": \"Streptomyces griseus\",\n",
    "    \"Alphaproteobacteria\": \"Rhizobium leguminosarum\",\n",
    "    \"Anaerolineae\": \"Anaerolinea thermophila\",\n",
    "    \"Aquificia\": \"Aquifex aeolicus\",\n",
    "    \"Archaeoglobi\": \"Archaeoglobus fulgidus\",\n",
    "    \"Armatimonadia\": \"Armatimonas rosea\",\n",
    "    \"Bacilli\": \"Bacillus subtilis\",\n",
    "    \"Bacteriovoracia\": \"Bacteriovorax stolpii\",\n",
    "    \"Bacteroidia\": \"Bacteroides fragilis\",\n",
    "    \"Balneolia\": \"Balneola vulgaris\",\n",
    "    \"Bdellovibrionia\": \"Bdellovibrio bacteriovorus\",\n",
    "    \"Betaproteobacteria\": \"Neisseria meningitidis\",\n",
    "    \"Blastocatellia\": \"Blastocatella fastidiosa\",\n",
    "    \"Caldilineae\": \"Caldilinea aerophila\",\n",
    "    \"Caldisericia\": \"Caldisericum exile\",\n",
    "    \"Chitinophagia\": \"Chitinophaga pinensis\",\n",
    "    \"Chlamydiia\": \"Chlamydia trachomatis\",\n",
    "    \"Chlorobiia\": \"Chlorobium tepidum\",\n",
    "    \"Chloroflexia\": \"Chloroflexus aurantiacus\",\n",
    "    \"Chrysiogenia\": \"Chrysiogenes arsenatis\",\n",
    "    \"Chthonomonadia\": \"Chthonomonas calidirosea\",\n",
    "    \"Clostridia\": \"Clostridium botulinum\",\n",
    "    \"Conexivisphaeria\": \"Conexibacter woesei\",\n",
    "    \"Coriobacteriia\": \"Coriobacterium glomerans\",\n",
    "    \"Cyanophyceae\": \"Synechocystis sp. PCC 6803\",\n",
    "    \"Cytophagia\": \"Cytophaga hutchinsonii\",\n",
    "    \"Deferribacteres\": \"Deferribacter thermophilus\",\n",
    "    \"Dehalococcoidia\": \"Dehalococcoides mccartyi\",\n",
    "    \"Deinococci\": \"Deinococcus radiodurans\",\n",
    "    \"Desulfarculia\": \"Desulfarculus baarsii\",\n",
    "    \"Desulfobaccia\": \"Desulfobacca acetoxidans\",\n",
    "    \"Desulfobacteria\": \"Desulfobacter hydrogenophilus\",\n",
    "    \"Desulfobulbia\": \"Desulfobulbus propionicus\",\n",
    "    \"Desulfomonilia\": \"Desulfomonile tiedjei\",\n",
    "    \"Desulfovibrionia\": \"Desulfovibrio vulgaris\",\n",
    "    \"Desulfurellia\": \"Desulfurella acetivorans\",\n",
    "    \"Desulfuromonadia\": \"Desulfuromonas acetoxidans\",\n",
    "    \"Dictyoglomia\": \"Dictyoglomus thermophilum\",\n",
    "    \"Elusimicrobia\": \"Elusimicrobium minutum\",\n",
    "    \"Endomicrobiia\": \"Endomicrobium proavitum\",\n",
    "    \"Epsilonproteobacteria\": \"Campylobacter jejuni\",\n",
    "    \"Erysipelotrichia\": \"Erysipelothrix rhusiopathiae\",\n",
    "    \"Fibrobacteria\": \"Fibrobacter succinogenes\",\n",
    "    \"Fimbriimonadia\": \"Fimbriimonas ginsengisoli\",\n",
    "    \"Flavobacteriia\": \"Flavobacterium johnsoniae\",\n",
    "    \"Fusobacteriia\": \"Fusobacterium nucleatum\",\n",
    "    \"Gammaproteobacteria\": \"Escherichia coli\",\n",
    "    \"Gemmatimonadia\": \"Gemmatimonas aurantiaca\",\n",
    "    \"Halobacteria\": \"Halobacterium salinarum\",\n",
    "    \"Holophagae\": \"Holophaga foetida\",\n",
    "    \"Hydrogenophilia\": \"Hydrogenophilus thermoluteolus\",\n",
    "    \"Ignavibacteria\": \"Ignavibacterium album\",\n",
    "    \"Kiritimatiellia\": \"Kiritimatiella glycovorans\",\n",
    "    \"Ktedonobacteria\": \"Ktedonobacter racemifer\",\n",
    "    \"Limnochordia\": \"Limnochorda pilosa\",\n",
    "    \"Methanobacteria\": \"Methanobacterium formicicum\",\n",
    "    \"Methanococci\": \"Methanococcus jannaschii\",\n",
    "    \"Methanomicrobia\": \"Methanosarcina barkeri\",\n",
    "    \"Methanonatronarchaeia\": \"Methanonatronarchaeum thermophilum\",\n",
    "    \"Methanopyri\": \"Methanopyrus kandleri\",\n",
    "    \"Methylacidiphilae\": \"Methylacidiphilum infernorum\",\n",
    "    \"Mollicutes\": \"Mycoplasma pneumoniae\",\n",
    "    \"Myxococcia\": \"Myxococcus xanthus\",\n",
    "    \"Nanobdellia\": \"Nanobdella aerobia\",\n",
    "    \"Negativicutes\": \"Veillonella parvula\",\n",
    "    \"Nitriliruptoria\": \"Nitriliruptor alkaliphilus\",\n",
    "    \"Nitrososphaeria\": \"Nitrosopumilus maritimus\",\n",
    "    \"Nitrospinia\": \"Nitrospina gracilis\",\n",
    "    \"Nitrospiria\": \"Nitrospira moscoviensis\",\n",
    "    \"Oligoflexia\": \"Oligoflexus tunisiensis\",\n",
    "    \"Opitutia\": \"Opitutus terrae\",\n",
    "    \"Phycisphaerae\": \"Phycisphaera mikurensis\",\n",
    "    \"Planctomycetia\": \"Planctomyces maris\",\n",
    "    \"Rhodothermia\": \"Rhodothermus marinus\",\n",
    "    \"Rubrobacteria\": \"Rubrobacter radiotolerans\",\n",
    "    \"Saprospiria\": \"Saprospira grandis\",\n",
    "    \"Sphingobacteriia\": \"Sphingobacterium spiritivorum\",\n",
    "    \"Spirochaetia\": \"Treponema pallidum\",\n",
    "    \"Synergistia\": \"Synergistes jonesii\",\n",
    "    \"Syntrophia\": \"Syntrophus aciditrophicus\",\n",
    "    \"Syntrophobacteria\": \"Syntrophobacter fumaroxidans\",\n",
    "    \"Tepidiformia\": \"Tepidiforma aggregans\",\n",
    "    \"Terriglobia\": \"Terriglobus roseus\",\n",
    "    \"Thermococci\": \"Thermococcus kodakarensis\",\n",
    "    \"Thermodesulfobacteria\": \"Thermodesulfobacterium commune\",\n",
    "    \"Thermodesulfobiia\": \"Thermodesulfobium narugense\",\n",
    "    \"Thermodesulfovibrionia\": \"Thermodesulfovibrio yellowstonii\",\n",
    "    \"Thermoflexia\": \"Thermoflexus hugenholtzii\",\n",
    "    \"Thermoleophilia\": \"Thermoleophilum album\",\n",
    "    \"Thermomicrobia\": \"Thermomicrobium roseum\",\n",
    "    \"Thermoprotei\": \"Sulfolobus solfataricus\",\n",
    "    \"Thermotogae\": \"Thermotoga maritima\",\n",
    "    \"Tichowtungiia\": None,\n",
    "    \"Tissierellia\": \"Tissierella praeacuta\",\n",
    "    \"Verrucomicrobiia\": \"Verrucomicrobium spinosum\",\n",
    "    \"Vicinamibacteria\": \"Vicinamibacter silvestris\",\n",
    "    \"Vulcanimicrobiia\": \"Vulcanimicrobium thermophilum\",\n",
    "    \"Zetaproteobacteria\": \"Mariprofundus ferrooxydans\"\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3a7b69ac-9e79-4156-8d31-4e644436758d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Step 1: Identify all classes and classify them as available or unavailable\n",
    "all_classes = set(taxonomy['Class'])\n",
    "available_classes = set(taxonomy[taxonomy['Species'].isin(representative_species.values())]['Class'])\n",
    "unavailable_classes = all_classes - available_classes\n",
    "\n",
    "# Step 2: Filter representative_species to only include available classes\n",
    "filtered_representative_species = {\n",
    "    class_: species for class_, species in representative_species.items()\n",
    "    if class_ in available_classes\n",
    "}\n",
    "\n",
    "# Step 3: Merge nr_genomes_dataset with taxonomy to get class information\n",
    "merged_genomes = nr_genomes_dataset.merge(taxonomy[['Species', 'Class']], on='Species').drop_duplicates()\n",
    "\n",
    "# Step 4: Create a dictionary for unavailable classes with one representative species\n",
    "unavailable_class_species = (\n",
    "    merged_genomes[\n",
    "        merged_genomes['Class'].isin(unavailable_classes) & (merged_genomes['Reference'] != '-')\n",
    "    ]\n",
    "    .drop_duplicates(subset=['Class'])\n",
    "    .set_index('Class')['Species']\n",
    "    .to_dict()\n",
    ")\n",
    "\n",
    "# Step 5: Combine filtered representative species with unavailable class species\n",
    "final_representative_species = {**filtered_representative_species, **unavailable_class_species}\n",
    "\n",
    "# Step 6: Handle remaining classes (if any) without references\n",
    "remaining_class_species = (\n",
    "    merged_genomes[\n",
    "        merged_genomes['Class'].isin(all_classes - set(final_representative_species.keys())) &\n",
    "        (merged_genomes['Reference'] == '-')\n",
    "    ]\n",
    "    .drop_duplicates(subset=['Class'])\n",
    "    .set_index('Class')['Species']\n",
    "    .to_dict()\n",
    ")\n",
    "\n",
    "# Step 7: Update the final representative species with remaining classes\n",
    "final_representative_species.update(remaining_class_species)\n",
    "\n",
    "# Step 8: Split genomes with and without references based on final species list\n",
    "final_species = set(final_representative_species.values())\n",
    "\n",
    "genomes_with_ref = genomes_dataset[\n",
    "    (genomes_dataset['Species'].isin(final_species)) & (genomes_dataset['Reference'] != '-')\n",
    "][['Organism', 'Species', 'Reference', 'Size (Mb)', 'ProteomeFile']].drop_duplicates(subset='Species', keep='first')\n",
    "\n",
    "genomes_without_ref = genomes_dataset[\n",
    "    (genomes_dataset['Species'].isin(final_species)) & (genomes_dataset['Reference'] == '-')\n",
    "][['Organism', 'Species', 'Reference', 'Size (Mb)', 'ProteomeFile']]\n",
    "\n",
    "# Exclude species already present in genomes_with_ref\n",
    "genomes_without_ref = genomes_without_ref[~genomes_without_ref['Species'].isin(genomes_with_ref['Species'])]\n",
    "genomes_without_ref.drop_duplicates(subset='Species', keep='first', inplace=True)\n",
    "\n",
    "# Step 9: Combine genomes with and without references for tree generation\n",
    "genomes_for_tree = pd.concat([genomes_with_ref, genomes_without_ref], ignore_index=True)\n",
    "\n",
    "# Final Output: `genomes_for_tree` contains the combined data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "16c542b3-79ac-4508-a46b-c348a8028515",
   "metadata": {},
   "outputs": [],
   "source": [
    "genomes_for_tree = genomes_for_tree.merge(taxonomy[['Species', 'Class']]).drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "47a7b9d8-9357-4d50-a02b-67ab4ddfe624",
   "metadata": {},
   "outputs": [],
   "source": [
    "# copy_files(genomes_for_tree['ProteomeFile'].unique(), '/Users/akshayonly/Work/04-Complex-I/Data/01-Sequence-Data/03-Proteomes', '/Users/akshayonly/Work/04-Complex-I/Data/06-Phylogeny-Data/Class-Tree/seqs')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "384e0939-f8e8-4654-ab66-530c06eb1543",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tree_info = pd.read_csv('/Users/akshayonly/Work/04-Complex-I/Data/06-Phylogeny-Data/Class-Tree/results/Genomes_summary_info.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "4b225149-0778-4514-8d04-5408e23f4aa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "genomes_for_tree['Node'] = genomes_for_tree['ProteomeFile'].apply(lambda x: x.replace('.faa', ''))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "1314a1c4-9787-4810-b240-83887c252832",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Organism</th>\n",
       "      <th>Species</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Size (Mb)</th>\n",
       "      <th>ProteomeFile</th>\n",
       "      <th>Class</th>\n",
       "      <th>Node</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Campylobacter jejuni</td>\n",
       "      <td>REFR</td>\n",
       "      <td>1.64148</td>\n",
       "      <td>GCA_000009085.1_ASM908v1_cds_proteins.faa</td>\n",
       "      <td>Epsilonproteobacteria</td>\n",
       "      <td>GCA_000009085.1_ASM908v1_cds_proteins</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>REFR</td>\n",
       "      <td>4.64165</td>\n",
       "      <td>GCA_000005845.2_ASM584v2_cds_proteins.faa</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>GCA_000005845.2_ASM584v2_cds_proteins</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Neisseria meningitidis</td>\n",
       "      <td>REPR</td>\n",
       "      <td>2.18133</td>\n",
       "      <td>GCA_022869645.1_ASM2286964v1_cds_proteins.faa</td>\n",
       "      <td>Betaproteobacteria</td>\n",
       "      <td>GCA_022869645.1_ASM2286964v1_cds_proteins</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Fusobacterium nucleatum</td>\n",
       "      <td>REPR</td>\n",
       "      <td>2.18010</td>\n",
       "      <td>GCF_003019295.1_ASM301929v1_cds_proteins.faa</td>\n",
       "      <td>Fusobacteriia</td>\n",
       "      <td>GCF_003019295.1_ASM301929v1_cds_proteins</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Chlorobium phaeobacteroides</td>\n",
       "      <td>REPR</td>\n",
       "      <td>3.13390</td>\n",
       "      <td>GCA_000015125.1_ASM1512v1_cds_proteins.faa</td>\n",
       "      <td>Chlorobiia</td>\n",
       "      <td>GCA_000015125.1_ASM1512v1_cds_proteins</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Organism                      Species Reference  Size (Mb)  \\\n",
       "0    Bacteria         Campylobacter jejuni      REFR    1.64148   \n",
       "33   Bacteria             Escherichia coli      REFR    4.64165   \n",
       "239  Bacteria       Neisseria meningitidis      REPR    2.18133   \n",
       "257  Bacteria      Fusobacterium nucleatum      REPR    2.18010   \n",
       "262  Bacteria  Chlorobium phaeobacteroides      REPR    3.13390   \n",
       "\n",
       "                                      ProteomeFile                  Class  \\\n",
       "0        GCA_000009085.1_ASM908v1_cds_proteins.faa  Epsilonproteobacteria   \n",
       "33       GCA_000005845.2_ASM584v2_cds_proteins.faa    Gammaproteobacteria   \n",
       "239  GCA_022869645.1_ASM2286964v1_cds_proteins.faa     Betaproteobacteria   \n",
       "257   GCF_003019295.1_ASM301929v1_cds_proteins.faa          Fusobacteriia   \n",
       "262     GCA_000015125.1_ASM1512v1_cds_proteins.faa             Chlorobiia   \n",
       "\n",
       "                                          Node  \n",
       "0        GCA_000009085.1_ASM908v1_cds_proteins  \n",
       "33       GCA_000005845.2_ASM584v2_cds_proteins  \n",
       "239  GCA_022869645.1_ASM2286964v1_cds_proteins  \n",
       "257   GCF_003019295.1_ASM301929v1_cds_proteins  \n",
       "262     GCA_000015125.1_ASM1512v1_cds_proteins  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genomes_for_tree.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "2fd25016-2764-4f5e-a57c-ff923da23151",
   "metadata": {},
   "outputs": [],
   "source": [
    "# genomes_for_tree[['Node', 'Class']].to_csv(\n",
    "#     '/Users/akshayonly/Work/04-Complex-I/Data/06-Phylogeny-Data/Class-Tree/class_labels.txt',\n",
    "#     sep=' ',\n",
    "#     index=False\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "988d8aff-60fd-4638-8330-2206507d2fa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "organism_colors = {'Bacteria': '#40407a', 'Archaea': '#218c74'}\n",
    "genomes_for_tree['OrganismColors'] = genomes_for_tree['Organism'].map(organism_colors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "37079ad8-2732-452d-892e-58b0a72716ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# genomes_for_tree[['Node', 'OrganismColors']].to_csv(\n",
    "#     '/Users/akshayonly/Work/04-Complex-I/Data/06-Phylogeny-Data/Class-Tree/organism.txt',\n",
    "#     sep=' ',\n",
    "#     index=False\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a1b3fe1-5bf7-4b5e-a317-9f34432d288f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
