{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "788ca799-b063-4f41-b7ed-1d2e53250fc7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from tqdm import tqdm\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import os\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4846d517-0142-4026-b718-afa0bd468584",
   "metadata": {},
   "outputs": [],
   "source": [
    "ndf_unfiltered = pd.read_csv('ndf_hmmer_search_hits.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4c90b604-ef6c-40b8-9ed1-1a4727c7bfe4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Subunit\n",
       "NDUFA9     445798\n",
       "NDUFA10      4285\n",
       "NDUFS4       2670\n",
       "NDUFA12      2513\n",
       "NDUFS6        919\n",
       "NDUFV3          6\n",
       "NDUFC1          2\n",
       "NDUFB10         1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ndf_unfiltered['Subunit'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1bb15f34-082b-4ded-a3cc-53b103dd5ee9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Boolean table for Nuo Complex I subunits variation data\n",
    "nuo_bool = pd.read_csv('nuo_bool_final.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4262f702-9f67-4baa-9a46-eef2c79c2866",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GenomeFile</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Superkingdom</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "      <th>Replicon</th>\n",
       "      <th>...</th>\n",
       "      <th>NuoH</th>\n",
       "      <th>NuoI</th>\n",
       "      <th>NuoJ</th>\n",
       "      <th>NuoK</th>\n",
       "      <th>NuoL</th>\n",
       "      <th>NuoM</th>\n",
       "      <th>NuoN</th>\n",
       "      <th>Variation</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_000009085.1_ASM908v1_genomic.fna</td>\n",
       "      <td>AL111168.1</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Campylobacterota</td>\n",
       "      <td>Epsilonproteobacteria</td>\n",
       "      <td>Campylobacterales</td>\n",
       "      <td>Campylobacteraceae</td>\n",
       "      <td>Campylobacter</td>\n",
       "      <td>Campylobacter jejuni</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>Nuo-like</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCF_001457695.1_NCTC11351_genomic.fna</td>\n",
       "      <td>NZ_LN831025.1</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Campylobacterota</td>\n",
       "      <td>Epsilonproteobacteria</td>\n",
       "      <td>Campylobacterales</td>\n",
       "      <td>Campylobacteraceae</td>\n",
       "      <td>Campylobacter</td>\n",
       "      <td>Campylobacter jejuni</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>Nuo-like</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCF_037680265.1_ASM3768026v1_genomic.fna</td>\n",
       "      <td>NZ_CP148213.1</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Campylobacterota</td>\n",
       "      <td>Epsilonproteobacteria</td>\n",
       "      <td>Campylobacterales</td>\n",
       "      <td>Campylobacteraceae</td>\n",
       "      <td>Campylobacter</td>\n",
       "      <td>Campylobacter jejuni</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>Nuo-like</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_033097125.1_ASM3309712v1_genomic.fna</td>\n",
       "      <td>AP028392.1</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Campylobacterota</td>\n",
       "      <td>Epsilonproteobacteria</td>\n",
       "      <td>Campylobacterales</td>\n",
       "      <td>Campylobacteraceae</td>\n",
       "      <td>Campylobacter</td>\n",
       "      <td>Campylobacter jejuni</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>Nuo-like</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_025232495.1_ASM2523249v1_genomic.fna</td>\n",
       "      <td>NZ_CP012237.1</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Campylobacterota</td>\n",
       "      <td>Epsilonproteobacteria</td>\n",
       "      <td>Campylobacterales</td>\n",
       "      <td>Campylobacteraceae</td>\n",
       "      <td>Campylobacter</td>\n",
       "      <td>Campylobacter jejuni</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>Nuo-like</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 GenomeFile      Accession Superkingdom  \\\n",
       "0      GCA_000009085.1_ASM908v1_genomic.fna     AL111168.1     Bacteria   \n",
       "1     GCF_001457695.1_NCTC11351_genomic.fna  NZ_LN831025.1     Bacteria   \n",
       "2  GCF_037680265.1_ASM3768026v1_genomic.fna  NZ_CP148213.1     Bacteria   \n",
       "3  GCA_033097125.1_ASM3309712v1_genomic.fna     AP028392.1     Bacteria   \n",
       "4  GCF_025232495.1_ASM2523249v1_genomic.fna  NZ_CP012237.1     Bacteria   \n",
       "\n",
       "             Phylum                  Class              Order  \\\n",
       "0  Campylobacterota  Epsilonproteobacteria  Campylobacterales   \n",
       "1  Campylobacterota  Epsilonproteobacteria  Campylobacterales   \n",
       "2  Campylobacterota  Epsilonproteobacteria  Campylobacterales   \n",
       "3  Campylobacterota  Epsilonproteobacteria  Campylobacterales   \n",
       "4  Campylobacterota  Epsilonproteobacteria  Campylobacterales   \n",
       "\n",
       "               Family          Genus               Species    Replicon  ...  \\\n",
       "0  Campylobacteraceae  Campylobacter  Campylobacter jejuni  Chromosome  ...   \n",
       "1  Campylobacteraceae  Campylobacter  Campylobacter jejuni  Chromosome  ...   \n",
       "2  Campylobacteraceae  Campylobacter  Campylobacter jejuni  Chromosome  ...   \n",
       "3  Campylobacteraceae  Campylobacter  Campylobacter jejuni  Chromosome  ...   \n",
       "4  Campylobacteraceae  Campylobacter  Campylobacter jejuni  Chromosome  ...   \n",
       "\n",
       "   NuoH  NuoI  NuoJ  NuoK  NuoL  NuoM  NuoN  Variation  Cluster  Strand  \n",
       "0  True  True  True  True  True  True  True   Nuo-like      NaN     NaN  \n",
       "1  True  True  True  True  True  True  True   Nuo-like      NaN     NaN  \n",
       "2  True  True  True  True  True  True  True   Nuo-like      NaN     NaN  \n",
       "3  True  True  True  True  True  True  True   Nuo-like      NaN     NaN  \n",
       "4  True  True  True  True  True  True  True   Nuo-like      NaN     NaN  \n",
       "\n",
       "[5 rows x 29 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nuo_bool.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "433dbd60-95b8-4dd3-84e0-f27fb018b1df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Replicon</th>\n",
       "      <th>ProteomeFile</th>\n",
       "      <th>Species</th>\n",
       "      <th>ProteinAccession</th>\n",
       "      <th>evalue</th>\n",
       "      <th>BitScore</th>\n",
       "      <th>Bias</th>\n",
       "      <th>SequenceDesc</th>\n",
       "      <th>Subunit</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>log10evalue</th>\n",
       "      <th>EstProtLength</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>250898</th>\n",
       "      <td>CP048788.1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GCA_012932215.1_ASM1293221v1_cds_proteins.faa</td>\n",
       "      <td>Roseobacter ponti</td>\n",
       "      <td>CP048788.1_3625</td>\n",
       "      <td>5.600000e-150</td>\n",
       "      <td>497.7</td>\n",
       "      <td>0.1</td>\n",
       "      <td># 3766663 # 3767646 # -1 # ID=1_3625;partial=0...</td>\n",
       "      <td>NDUFA9</td>\n",
       "      <td>3766663</td>\n",
       "      <td>3767646</td>\n",
       "      <td>-149.251812</td>\n",
       "      <td>328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313207</th>\n",
       "      <td>CP090021.1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GCA_021391315.1_ASM2139131v1_cds_proteins.faa</td>\n",
       "      <td>Cereibacter azotoformans</td>\n",
       "      <td>CP090021.1_2799</td>\n",
       "      <td>3.300000e-148</td>\n",
       "      <td>492.1</td>\n",
       "      <td>0.8</td>\n",
       "      <td># 2825453 # 2826439 # -1 # ID=1_2799;partial=0...</td>\n",
       "      <td>NDUFA9</td>\n",
       "      <td>2825453</td>\n",
       "      <td>2826439</td>\n",
       "      <td>-147.481486</td>\n",
       "      <td>329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30944</th>\n",
       "      <td>JAGFXX010000001.1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GCA_017599285.1_ASM1759928v1_cds_proteins.faa</td>\n",
       "      <td>Cereibacter azotoformans</td>\n",
       "      <td>JAGFXX010000001.1_2562</td>\n",
       "      <td>2.400000e-148</td>\n",
       "      <td>492.1</td>\n",
       "      <td>0.8</td>\n",
       "      <td># 2586504 # 2587490 # 1 # ID=1_2562;partial=00...</td>\n",
       "      <td>NDUFA9</td>\n",
       "      <td>2586504</td>\n",
       "      <td>2587490</td>\n",
       "      <td>-147.619789</td>\n",
       "      <td>329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246241</th>\n",
       "      <td>CP089965.1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GCA_022227035.1_ASM2222703v1_cds_proteins.faa</td>\n",
       "      <td>Cereibacter azotoformans</td>\n",
       "      <td>CP089965.1_2823</td>\n",
       "      <td>3.600000e-148</td>\n",
       "      <td>492.1</td>\n",
       "      <td>0.8</td>\n",
       "      <td># 2843382 # 2844368 # -1 # ID=1_2823;partial=0...</td>\n",
       "      <td>NDUFA9</td>\n",
       "      <td>2843382</td>\n",
       "      <td>2844368</td>\n",
       "      <td>-147.443697</td>\n",
       "      <td>329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183734</th>\n",
       "      <td>NZ_CP033446.1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GCF_003846385.1_ASM384638v1_cds_proteins.faa</td>\n",
       "      <td>Cereibacter sphaeroides</td>\n",
       "      <td>NZ_CP033446.1_261</td>\n",
       "      <td>6.300000e-147</td>\n",
       "      <td>487.8</td>\n",
       "      <td>0.1</td>\n",
       "      <td># 272601 # 273587 # 1 # ID=1_261;partial=00;st...</td>\n",
       "      <td>NDUFA9</td>\n",
       "      <td>272601</td>\n",
       "      <td>273587</td>\n",
       "      <td>-146.200659</td>\n",
       "      <td>329</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Accession    Replicon  \\\n",
       "250898         CP048788.1  Chromosome   \n",
       "313207         CP090021.1  Chromosome   \n",
       "30944   JAGFXX010000001.1  Chromosome   \n",
       "246241         CP089965.1  Chromosome   \n",
       "183734      NZ_CP033446.1  Chromosome   \n",
       "\n",
       "                                         ProteomeFile  \\\n",
       "250898  GCA_012932215.1_ASM1293221v1_cds_proteins.faa   \n",
       "313207  GCA_021391315.1_ASM2139131v1_cds_proteins.faa   \n",
       "30944   GCA_017599285.1_ASM1759928v1_cds_proteins.faa   \n",
       "246241  GCA_022227035.1_ASM2222703v1_cds_proteins.faa   \n",
       "183734   GCF_003846385.1_ASM384638v1_cds_proteins.faa   \n",
       "\n",
       "                         Species        ProteinAccession         evalue  \\\n",
       "250898         Roseobacter ponti         CP048788.1_3625  5.600000e-150   \n",
       "313207  Cereibacter azotoformans         CP090021.1_2799  3.300000e-148   \n",
       "30944   Cereibacter azotoformans  JAGFXX010000001.1_2562  2.400000e-148   \n",
       "246241  Cereibacter azotoformans         CP089965.1_2823  3.600000e-148   \n",
       "183734   Cereibacter sphaeroides       NZ_CP033446.1_261  6.300000e-147   \n",
       "\n",
       "        BitScore  Bias                                       SequenceDesc  \\\n",
       "250898     497.7   0.1  # 3766663 # 3767646 # -1 # ID=1_3625;partial=0...   \n",
       "313207     492.1   0.8  # 2825453 # 2826439 # -1 # ID=1_2799;partial=0...   \n",
       "30944      492.1   0.8  # 2586504 # 2587490 # 1 # ID=1_2562;partial=00...   \n",
       "246241     492.1   0.8  # 2843382 # 2844368 # -1 # ID=1_2823;partial=0...   \n",
       "183734     487.8   0.1  # 272601 # 273587 # 1 # ID=1_261;partial=00;st...   \n",
       "\n",
       "       Subunit    Start      End  log10evalue  EstProtLength  \n",
       "250898  NDUFA9  3766663  3767646  -149.251812            328  \n",
       "313207  NDUFA9  2825453  2826439  -147.481486            329  \n",
       "30944   NDUFA9  2586504  2587490  -147.619789            329  \n",
       "246241  NDUFA9  2843382  2844368  -147.443697            329  \n",
       "183734  NDUFA9   272601   273587  -146.200659            329  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ndf_unfiltered.sort_values('BitScore', ascending=False).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c47599f8-f59a-4009-9cb9-bed30808d232",
   "metadata": {},
   "outputs": [],
   "source": [
    "colors = [\"#a5b1c2\", \"#f7b731\", \"#20bf6b\", \"#45aaf2\", \"#45aaf2\", \"#3867d6\", \"#a55eea\", \"#0fb9b1\", '#4b6584', \"#eb3b5a\"]\n",
    "labels = ['Nuo-Partial', 'Nuo13', 'Nuo14', 'Nuo-like', 'Nuo14-EFG', 'Nuo13-EFG', 'Nuo14-EF', 'Nuo12', 'Existing Annotation', 'Absent']\n",
    "\n",
    "complex_colors = dict(zip(labels, colors))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e4e340b1-a50d-476f-bf64-d20ee96f9ed5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Subunit\n",
       "NDUFA9     445798\n",
       "NDUFA10      4285\n",
       "NDUFS4       2670\n",
       "NDUFA12      2513\n",
       "NDUFS6        919\n",
       "NDUFV3          6\n",
       "NDUFC1          2\n",
       "NDUFB10         1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ndf_unfiltered['Subunit'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "23f9094f-60b4-4f9a-8917-03fe47132e24",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved: acc_kde_plots/NDUFA10_kde.png\n",
      "Saved: acc_kde_plots/NDUFA12_kde.png\n",
      "Saved: acc_kde_plots/NDUFA9_kde.png\n",
      "Saved: acc_kde_plots/NDUFB10_kde.png\n",
      "Saved: acc_kde_plots/NDUFC1_kde.png\n",
      "Saved: acc_kde_plots/NDUFS4_kde.png\n",
      "Saved: acc_kde_plots/NDUFS6_kde.png\n",
      "Saved: acc_kde_plots/NDUFV3_kde.png\n"
     ]
    }
   ],
   "source": [
    "# List of subunits to analyze\n",
    "subunits = sorted(ndf_unfiltered['Subunit'].unique())\n",
    "\n",
    "# Directory to save plots\n",
    "output_dir = \"acc_kde_plots\"\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "for subunit in subunits:\n",
    "    # Filter HMMER data for the current subunit\n",
    "    filtered_hmmer_data = ndf_unfiltered.loc[\n",
    "        ndf_unfiltered['Subunit'] == subunit, \n",
    "        ['Accession', 'log10evalue']\n",
    "    ].drop_duplicates()\n",
    "\n",
    "    # Extract unique accession-variation mappings\n",
    "    unique_variations = nuo_bool[['Accession', 'Variation']].drop_duplicates()\n",
    "\n",
    "    # Identify missing accessions and assign 'Absent' variation\n",
    "    missing_accessions = filtered_hmmer_data.loc[\n",
    "        ~filtered_hmmer_data['Accession'].isin(unique_variations['Accession']), \n",
    "        ['Accession']\n",
    "    ].assign(Variation='Absent')\n",
    "\n",
    "    # Combine variation data\n",
    "    variation_status = pd.concat([unique_variations, missing_accessions], ignore_index=True).drop_duplicates()\n",
    "\n",
    "    # Merge filtered HMMER data with variation status\n",
    "    plot_data = filtered_hmmer_data.merge(variation_status, on='Accession', how='left')\n",
    "\n",
    "    # Generate KDE plot only if there is valid data\n",
    "    if 'Variation' in plot_data.columns and not plot_data.empty:\n",
    "        plt.figure(figsize=(8, 6))\n",
    "        sns.kdeplot(data=plot_data, x='log10evalue', hue='Variation', palette=complex_colors)\n",
    "        plt.title(f\"KDE Plot for {subunit}\")\n",
    "        plt.xlabel(\"log10(E-value)\")\n",
    "        plt.ylabel(\"Density\")\n",
    "\n",
    "        # Save the plot\n",
    "        plot_filename = os.path.join(output_dir, f\"{subunit}_kde.png\")\n",
    "        plt.savefig(plot_filename, dpi=300)\n",
    "        plt.close()  # Close the plot to avoid overlap in the loop\n",
    "        print(f\"Saved: {plot_filename}\")\n",
    "    else:\n",
    "        print(f\"Warning: No valid data for {subunit}, skipping plot.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5e924bb3-2823-495b-b2e5-b476f7326f9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Based on the KDE distibution, e-value cutoffs are selected\n",
    "\n",
    "E_VALUE_CUTOFF = {'NDUFS4' : -25, 'NDUFA12' : -35, 'NDUFA9' : -75}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "168716cb-075d-4cfa-8adf-d63e1d81552b",
   "metadata": {},
   "outputs": [],
   "source": [
    "selected_ndf = ndf_unfiltered[ndf_unfiltered['Subunit'].isin(E_VALUE_CUTOFF.keys())]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f73b8b8f-a1a2-4089-ac72-3daa2e33263a",
   "metadata": {},
   "outputs": [],
   "source": [
    "selected_ndf = selected_ndf[selected_ndf['log10evalue'] <= selected_ndf['Subunit'].map(E_VALUE_CUTOFF).fillna(float('inf'))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "9af050ae-3508-440a-a83f-064612e45d00",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Index: 9643 entries, 110 to 456030\n",
      "Data columns (total 14 columns):\n",
      " #   Column            Non-Null Count  Dtype  \n",
      "---  ------            --------------  -----  \n",
      " 0   Accession         9643 non-null   object \n",
      " 1   Replicon          9643 non-null   object \n",
      " 2   ProteomeFile      9643 non-null   object \n",
      " 3   Species           9643 non-null   object \n",
      " 4   ProteinAccession  9643 non-null   object \n",
      " 5   evalue            9643 non-null   float64\n",
      " 6   BitScore          9643 non-null   float64\n",
      " 7   Bias              9643 non-null   float64\n",
      " 8   SequenceDesc      9643 non-null   object \n",
      " 9   Subunit           9643 non-null   object \n",
      " 10  Start             9643 non-null   int64  \n",
      " 11  End               9643 non-null   int64  \n",
      " 12  log10evalue       9643 non-null   float64\n",
      " 13  EstProtLength     9643 non-null   int64  \n",
      "dtypes: float64(4), int64(3), object(7)\n",
      "memory usage: 1.1+ MB\n"
     ]
    }
   ],
   "source": [
    "selected_ndf.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "80c56503-214c-4646-a11f-ef4a544aa612",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set output directory\n",
    "proteomes_dir = '/proteomes'\n",
    "\n",
    "output_dir = '/acc_seqs'\n",
    "\n",
    "for subunit in sorted(selected_ndf['Subunit'].unique()):\n",
    "    selected_selected_ndf = selected_ndf[selected_ndf['Subunit'] == subunit]\n",
    "    \n",
    "    # File to store the combined sequences\n",
    "    output_fasta = f\"{output_dir}/{subunit}_filtered_hits.fasta\"\n",
    "    \n",
    "    # Create a mapping of proteome files to protein accessions\n",
    "    proteome_accessions = selected_selected_ndf.groupby('ProteomeFile')['ProteinAccession'].unique().to_dict()\n",
    "    \n",
    "    # Open the output file once\n",
    "    with open(output_fasta, 'w') as output_handle:\n",
    "        # Process each proteome file only once\n",
    "        for proteome_file, accessions in proteome_accessions.items():\n",
    "            proteome_file_fp = os.path.join(proteomes_dir, proteome_file)\n",
    "            try:\n",
    "                # Parse the proteome file and extract sequences for needed accessions\n",
    "                needed_accessions = set(accessions)\n",
    "                sequences_found = 0\n",
    "                \n",
    "                for record in SeqIO.parse(proteome_file_fp, 'fasta'):\n",
    "                    if record.id in needed_accessions:\n",
    "                        # Clean the sequence\n",
    "                        protein_sequence = str(record.seq).replace('*', '')\n",
    "                        \n",
    "                        # Create a SeqRecord with the cleaned sequence\n",
    "                        seq_record = SeqRecord(\n",
    "                            seq=protein_sequence,\n",
    "                            id=record.id,\n",
    "                            description=record.description\n",
    "                        )\n",
    "                        \n",
    "                        # Write the SeqRecord to the output FASTA file\n",
    "                        SeqIO.write(seq_record, output_handle, 'fasta')\n",
    "                        \n",
    "                        sequences_found += 1\n",
    "                        needed_accessions.remove(record.id)\n",
    "                        \n",
    "                        # Break if all needed accessions have been found\n",
    "                        if not needed_accessions:\n",
    "                            break\n",
    "                if sequences_found == 0:\n",
    "                    print(f\"No sequences found in {proteome_file_fp} for the specified accessions.\")\n",
    "            except FileNotFoundError:\n",
    "                print(f\"File not found: {proteome_file_fp}\")\n",
    "            except Exception as e:\n",
    "                print(f\"An error occurred with file {proteome_file_fp}: {e}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b6885b1-9f6e-4f1e-9a85-88efd835e223",
   "metadata": {},
   "source": [
    "These sequences are further used to build phylogenetic trees"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
