{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ee8b4ec2-df40-45b2-9ce9-8e934b57ed3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pytaxonkit\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "736678b2-50a9-4da5-b867-8b359c66ac03",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_prokaryote_taxon(data):\n",
    "    \"\"\"\n",
    "    Processes a DataFrame to extract and structure taxonomic information using PyTaxonKit.\n",
    "    \n",
    "    Parameters:\n",
    "    - data: DataFrame with 'TaxID' column.\n",
    "    \n",
    "    Returns:\n",
    "    - DataFrame with taxonomic information columns added.\n",
    "    \"\"\"\n",
    "    # Get taxonomic lineage using PyTaxonKit\n",
    "    data_taxonkit = pytaxonkit.lineage(data[\"TaxID\"].unique(), formatstr=\"{s}\", threads=14)\n",
    "    # Preprocess lineage information for easier data manipulation\n",
    "    taxonomic_info = []\n",
    "    for index, row in tqdm(data_taxonkit.iterrows(), total=data_taxonkit.shape[0], desc=\"Processing taxonomic information\"):\n",
    "        lineage_id = row['LineageTaxIDs']\n",
    "        lineage_name = row['Lineage']\n",
    "        lineage_ranks = row['FullLineageRanks'].split(';')\n",
    "        full_lineage = row['FullLineage'].split(';')\n",
    "        taxonomy = dict(zip(lineage_ranks, full_lineage))\n",
    "        taxonomic_info.append({\n",
    "            'TaxID': row['TaxID'],\n",
    "            'Organism': full_lineage[1] if len(full_lineage) > 1 else None,\n",
    "            'Lineage' : lineage_name,\n",
    "            'LineageTaxIDs' : lineage_id,\n",
    "            'Clade': taxonomy.get('clade'),\n",
    "            'Order': taxonomy.get('order'),\n",
    "            'Class': taxonomy.get('class'),\n",
    "            'Phylum': taxonomy.get('phylum'),\n",
    "            'Genus': taxonomy.get('genus')\n",
    "        })\n",
    "    taxonomic_df = pd.DataFrame(taxonomic_info)\n",
    "    return pd.merge(data, taxonomic_df, on='TaxID', how='left').drop_duplicates()\n",
    "\n",
    "def load_and_prepare_prokaryotes(main_dir):\n",
    "    \"\"\"\n",
    "    Loads prokaryotes data and prepares it for processing.\n",
    "    \n",
    "    Parameters:\n",
    "    - main_dir: Directory containing the prokaryotes.txt file.\n",
    "    \n",
    "    Returns:\n",
    "    - DataFrame of prokaryotes with essential information and FTP paths.\n",
    "    \"\"\"\n",
    "    prokaryotes_path = os.path.join(main_dir, \"prokaryotes.txt\")\n",
    "    prokaryotes = pd.read_csv(prokaryotes_path, sep='\\t', low_memory=False)\n",
    "    # prokaryotes['FTP Path'].replace('-', np.nan, inplace=True)\n",
    "    prokaryotes['FTP Path'] = prokaryotes['FTP Path'].replace('-', np.nan)\n",
    "    prokaryotes.dropna(subset=['FTP Path'], inplace=True)\n",
    "    \n",
    "    # Add columns for genome and CDS file paths\n",
    "    ftp_to_path = lambda x: f\"{x}/{x.split('/')[-1]}\"\n",
    "    prokaryotes[\"GenomePath\"] = prokaryotes[\"FTP Path\"].apply(lambda x: ftp_to_path(x) + \"_genomic.fna.gz\")\n",
    "    prokaryotes[\"CDSPath\"] = prokaryotes[\"FTP Path\"].apply(lambda x: ftp_to_path(x) + \"_cds_from_genomic.fna.gz\")\n",
    "\n",
    "    ftp_to_path = lambda x: f\"{x.split('/')[-1]}\"\n",
    "    prokaryotes[\"GenomeFile\"] = prokaryotes[\"FTP Path\"].apply(lambda x: ftp_to_path(x) + \"_genomic.fna\")\n",
    "    prokaryotes[\"CDSFile\"] = prokaryotes[\"FTP Path\"].apply(lambda x: ftp_to_path(x) + \"_cds_from_genomic.fna\")\n",
    "    \n",
    "    return prokaryotes\n",
    "\n",
    "def filter_and_process_taxon(prokaryotes):\n",
    "    \"\"\"\n",
    "    Filters prokaryotes by completeness status, processes taxonomic information,\n",
    "    and merges the taxonomic information back into the main DataFrame.\n",
    "    \n",
    "    Parameters:\n",
    "    - prokaryotes: DataFrame of prokaryotes.\n",
    "    \n",
    "    Returns:\n",
    "    - DataFrame with processed taxonomic information.\n",
    "    \"\"\"\n",
    "    # Filter by Reference and Status\n",
    "    # reference = prokaryotes[prokaryotes['Reference'].isin(['REFR', 'REPR'])]\n",
    "    complete = prokaryotes[prokaryotes['Status'].isin(['Complete Genome', 'Chromosome', 'Complete', 'Chromosome(s)'])]\n",
    "\n",
    "    # Process taxonomic information\n",
    "    complete_taxon = process_prokaryote_taxon(complete)\n",
    "    # reference_taxon = process_prokaryote_taxon(reference)\n",
    "    \n",
    "    # Combine processed data\n",
    "    # combined_taxon = pd.concat([complete_taxon, reference_taxon]).drop_duplicates()\n",
    "    return complete_taxon\n",
    "\n",
    "def save_output(prokaryotes_genomes, dir_path):\n",
    "    \"\"\"\n",
    "    Saves the processed prokaryotes genomes information to CSV and FTP path files.\n",
    "    \n",
    "    Parameters:\n",
    "    - prokaryotes_genomes: DataFrame with prokaryotes genomes information.\n",
    "    - dir_path: Directory path to save the output files.\n",
    "    \"\"\"\n",
    "    # Save to CSV\n",
    "    prokaryotes_genomes.to_csv(os.path.join(dir_path, \"prokaryotes_processed.csv\"), index=False)\n",
    "    \n",
    "    # Save FTP paths\n",
    "    with open(os.path.join(dir_path, 'ncbi_cds_ftp.txt'), 'w') as f:\n",
    "        for ftp in prokaryotes_genomes['CDSPath']:\n",
    "            f.write(f\"{ftp}\\n\")\n",
    "    \n",
    "    with open(os.path.join(dir_path, 'ncbi_genomes_ftp.txt'), 'w') as f:\n",
    "        for ftp in prokaryotes_genomes['GenomePath']:\n",
    "            f.write(f\"{ftp}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d714086d-46bf-4ea5-8da6-88cb27271db9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing taxonomic information: 100%|████████████████████████████████████████| 15591/15591 [00:00<00:00, 75819.66it/s]\n"
     ]
    }
   ],
   "source": [
    "main_dir = \"/Users/akshayonly/Work/Updated/Data/01/\"\n",
    "prokaryotes = load_and_prepare_prokaryotes(main_dir)\n",
    "prokaryotes_genomes = filter_and_process_taxon(prokaryotes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f864e2f0-1c2c-4140-964f-26dba92f43d1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 45976 entries, 0 to 45975\n",
      "Data columns (total 35 columns):\n",
      " #   Column                Non-Null Count  Dtype  \n",
      "---  ------                --------------  -----  \n",
      " 0   #Organism/Name        45976 non-null  object \n",
      " 1   TaxID                 45976 non-null  int64  \n",
      " 2   BioProject Accession  45976 non-null  object \n",
      " 3   BioProject ID         45976 non-null  int64  \n",
      " 4   Group                 45976 non-null  object \n",
      " 5   SubGroup              45976 non-null  object \n",
      " 6   Size (Mb)             45976 non-null  float64\n",
      " 7   GC%                   45976 non-null  object \n",
      " 8   Replicons             45976 non-null  object \n",
      " 9   WGS                   45976 non-null  object \n",
      " 10  Scaffolds             45976 non-null  object \n",
      " 11  Genes                 45976 non-null  object \n",
      " 12  Proteins              45976 non-null  object \n",
      " 13  Release Date          45976 non-null  object \n",
      " 14  Modify Date           45976 non-null  object \n",
      " 15  Status                45976 non-null  object \n",
      " 16  Center                45932 non-null  object \n",
      " 17  BioSample Accession   45976 non-null  object \n",
      " 18  Assembly Accession    45976 non-null  object \n",
      " 19  Reference             45976 non-null  object \n",
      " 20  FTP Path              45976 non-null  object \n",
      " 21  Pubmed ID             45976 non-null  object \n",
      " 22  Strain                45976 non-null  object \n",
      " 23  GenomePath            45976 non-null  object \n",
      " 24  CDSPath               45976 non-null  object \n",
      " 25  GenomeFile            45976 non-null  object \n",
      " 26  CDSFile               45976 non-null  object \n",
      " 27  Organism              45976 non-null  object \n",
      " 28  Lineage               45976 non-null  object \n",
      " 29  LineageTaxIDs         45976 non-null  int64  \n",
      " 30  Clade                 18929 non-null  object \n",
      " 31  Order                 45307 non-null  object \n",
      " 32  Class                 45058 non-null  object \n",
      " 33  Phylum                45917 non-null  object \n",
      " 34  Genus                 44918 non-null  object \n",
      "dtypes: float64(1), int64(3), object(31)\n",
      "memory usage: 12.3+ MB\n"
     ]
    }
   ],
   "source": [
    "prokaryotes_genomes.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f335f152-9981-4fc6-a230-3a0e5ae835ed",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11669"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prokaryotes_genomes['LineageTaxIDs'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c6f1e082",
   "metadata": {},
   "outputs": [],
   "source": [
    "# First filter out rows with NaN in 'Phylum'\n",
    "prokaryotes_genomes = prokaryotes_genomes[prokaryotes_genomes['Phylum'].notna()]\n",
    "\n",
    "# Then filter out rows where 'Phylum' starts with 'Candidatus' or 'candidate'\n",
    "prokaryotes_genomes = prokaryotes_genomes[\n",
    "    ~prokaryotes_genomes['Phylum'].str.startswith('Candidatus') & \n",
    "    ~prokaryotes_genomes['Phylum'].str.startswith('candidate')\n",
    "]\n",
    "\n",
    "# Get phylum counts and filter phyla with counts >= 2\n",
    "phylum_counts = prokaryotes_genomes['Phylum'].value_counts()\n",
    "selected_phyla = phylum_counts[phylum_counts >= 2].index\n",
    "\n",
    "# Filter the genomes based on the selected phyla\n",
    "prokaryotes_genomes = prokaryotes_genomes[prokaryotes_genomes['Phylum'].isin(selected_phyla)]\n",
    "\n",
    "# Filter the genomes which contains Genus information\n",
    "prokaryotes_genomes = prokaryotes_genomes[~prokaryotes_genomes['Genus'].isnull()]\n",
    "\n",
    "prokaryotes_genomes = prokaryotes_genomes[~prokaryotes_genomes['Genus'].str.startswith('Candidatus')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "74ff6413",
   "metadata": {},
   "outputs": [],
   "source": [
    "genus = prokaryotes_genomes['Genus'].value_counts().reset_index()\n",
    "\n",
    "# Group by 'Genus' and filter groups with at least 2 unique 'LineageTaxIDs'\n",
    "genus_to_keep = (\n",
    "    prokaryotes_genomes.groupby('Genus')\n",
    "    .filter(lambda x: x['LineageTaxIDs'].nunique() >= 2)['Genus']\n",
    "    .unique()\n",
    ")\n",
    "\n",
    "# Filter the original DataFrame to keep only the selected genera\n",
    "prokaryotes_genomes = prokaryotes_genomes[prokaryotes_genomes['Genus'].isin(genus_to_keep)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d10f8bfe",
   "metadata": {},
   "outputs": [],
   "source": [
    "prokaryotes_genomes = prokaryotes_genomes.drop_duplicates().reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ab7da602",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_output(prokaryotes_genomes, main_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "bd56fdae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#Organism/Name</th>\n",
       "      <th>TaxID</th>\n",
       "      <th>BioProject Accession</th>\n",
       "      <th>BioProject ID</th>\n",
       "      <th>Group</th>\n",
       "      <th>SubGroup</th>\n",
       "      <th>Size (Mb)</th>\n",
       "      <th>GC%</th>\n",
       "      <th>Replicons</th>\n",
       "      <th>WGS</th>\n",
       "      <th>...</th>\n",
       "      <th>GenomeFile</th>\n",
       "      <th>CDSFile</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Lineage</th>\n",
       "      <th>LineageTaxIDs</th>\n",
       "      <th>Clade</th>\n",
       "      <th>Order</th>\n",
       "      <th>Class</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Genus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Campylobacter jejuni subsp. jejuni NCTC 11168 ...</td>\n",
       "      <td>192222</td>\n",
       "      <td>PRJNA8</td>\n",
       "      <td>8</td>\n",
       "      <td>Campylobacterota</td>\n",
       "      <td>Epsilonproteobacteria</td>\n",
       "      <td>1.64148</td>\n",
       "      <td>30.5</td>\n",
       "      <td>chromosome:NC_002163.1/AL111168.1</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>GCA_000009085.1_ASM908v1_genomic.fna</td>\n",
       "      <td>GCA_000009085.1_ASM908v1_cds_from_genomic.fna</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Campylobacter jejuni</td>\n",
       "      <td>197</td>\n",
       "      <td>None</td>\n",
       "      <td>Campylobacterales</td>\n",
       "      <td>Epsilonproteobacteria</td>\n",
       "      <td>Campylobacterota</td>\n",
       "      <td>Campylobacter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Pseudomonas fluorescens</td>\n",
       "      <td>294</td>\n",
       "      <td>PRJEB22404</td>\n",
       "      <td>401622</td>\n",
       "      <td>Pseudomonadota</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>6.51155</td>\n",
       "      <td>60</td>\n",
       "      <td>chromosome I:NZ_LT907842.1/LT907842.1</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>GCA_900215245.1_IMG-taxon_2617270901_annotated...</td>\n",
       "      <td>GCA_900215245.1_IMG-taxon_2617270901_annotated...</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Pseudomonas fluorescens</td>\n",
       "      <td>294</td>\n",
       "      <td>None</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadota</td>\n",
       "      <td>Pseudomonas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Xanthomonas campestris pv. raphani</td>\n",
       "      <td>359385</td>\n",
       "      <td>PRJNA641237</td>\n",
       "      <td>641237</td>\n",
       "      <td>Pseudomonadota</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>4.94204</td>\n",
       "      <td>65.3</td>\n",
       "      <td>chromosome:NZ_CP058243.1/CP058243.1</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>GCA_013388375.1_ASM1338837v1_genomic.fna</td>\n",
       "      <td>GCA_013388375.1_ASM1338837v1_cds_from_genomic.fna</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Xanthomonas campestris</td>\n",
       "      <td>339</td>\n",
       "      <td>None</td>\n",
       "      <td>Lysobacterales</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadota</td>\n",
       "      <td>Xanthomonas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ty...</td>\n",
       "      <td>99287</td>\n",
       "      <td>PRJNA241</td>\n",
       "      <td>241</td>\n",
       "      <td>Pseudomonadota</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>4.95138</td>\n",
       "      <td>52.2171</td>\n",
       "      <td>chromosome:NC_003197.2/AE006468.2; plasmid pSL...</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>GCA_000006945.2_ASM694v2_genomic.fna</td>\n",
       "      <td>GCA_000006945.2_ASM694v2_cds_from_genomic.fna</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>28901</td>\n",
       "      <td>None</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadota</td>\n",
       "      <td>Salmonella</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Yersinia pestis A1122</td>\n",
       "      <td>1035377</td>\n",
       "      <td>PRJNA67155</td>\n",
       "      <td>67155</td>\n",
       "      <td>Pseudomonadota</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>4.65841</td>\n",
       "      <td>47.6472</td>\n",
       "      <td>chromosome:CP002956.1; plasmid unnamed:CP00295...</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>GCA_000222975.1_ASM22297v1_genomic.fna</td>\n",
       "      <td>GCA_000222975.1_ASM22297v1_cds_from_genomic.fna</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Yersinia pestis</td>\n",
       "      <td>632</td>\n",
       "      <td>None</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadota</td>\n",
       "      <td>Yersinia</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      #Organism/Name    TaxID  \\\n",
       "0  Campylobacter jejuni subsp. jejuni NCTC 11168 ...   192222   \n",
       "1                            Pseudomonas fluorescens      294   \n",
       "2                 Xanthomonas campestris pv. raphani   359385   \n",
       "3  Salmonella enterica subsp. enterica serovar Ty...    99287   \n",
       "4                              Yersinia pestis A1122  1035377   \n",
       "\n",
       "  BioProject Accession  BioProject ID             Group  \\\n",
       "0               PRJNA8              8  Campylobacterota   \n",
       "1           PRJEB22404         401622    Pseudomonadota   \n",
       "2          PRJNA641237         641237    Pseudomonadota   \n",
       "3             PRJNA241            241    Pseudomonadota   \n",
       "4           PRJNA67155          67155    Pseudomonadota   \n",
       "\n",
       "                SubGroup  Size (Mb)      GC%  \\\n",
       "0  Epsilonproteobacteria    1.64148     30.5   \n",
       "1    Gammaproteobacteria    6.51155       60   \n",
       "2    Gammaproteobacteria    4.94204     65.3   \n",
       "3    Gammaproteobacteria    4.95138  52.2171   \n",
       "4    Gammaproteobacteria    4.65841  47.6472   \n",
       "\n",
       "                                           Replicons WGS  ...  \\\n",
       "0                  chromosome:NC_002163.1/AL111168.1   -  ...   \n",
       "1              chromosome I:NZ_LT907842.1/LT907842.1   -  ...   \n",
       "2                chromosome:NZ_CP058243.1/CP058243.1   -  ...   \n",
       "3  chromosome:NC_003197.2/AE006468.2; plasmid pSL...   -  ...   \n",
       "4  chromosome:CP002956.1; plasmid unnamed:CP00295...   -  ...   \n",
       "\n",
       "                                          GenomeFile  \\\n",
       "0               GCA_000009085.1_ASM908v1_genomic.fna   \n",
       "1  GCA_900215245.1_IMG-taxon_2617270901_annotated...   \n",
       "2           GCA_013388375.1_ASM1338837v1_genomic.fna   \n",
       "3               GCA_000006945.2_ASM694v2_genomic.fna   \n",
       "4             GCA_000222975.1_ASM22297v1_genomic.fna   \n",
       "\n",
       "                                             CDSFile  Organism  \\\n",
       "0      GCA_000009085.1_ASM908v1_cds_from_genomic.fna  Bacteria   \n",
       "1  GCA_900215245.1_IMG-taxon_2617270901_annotated...  Bacteria   \n",
       "2  GCA_013388375.1_ASM1338837v1_cds_from_genomic.fna  Bacteria   \n",
       "3      GCA_000006945.2_ASM694v2_cds_from_genomic.fna  Bacteria   \n",
       "4    GCA_000222975.1_ASM22297v1_cds_from_genomic.fna  Bacteria   \n",
       "\n",
       "                   Lineage LineageTaxIDs Clade              Order  \\\n",
       "0     Campylobacter jejuni           197  None  Campylobacterales   \n",
       "1  Pseudomonas fluorescens           294  None    Pseudomonadales   \n",
       "2   Xanthomonas campestris           339  None     Lysobacterales   \n",
       "3      Salmonella enterica         28901  None   Enterobacterales   \n",
       "4          Yersinia pestis           632  None   Enterobacterales   \n",
       "\n",
       "                   Class            Phylum          Genus  \n",
       "0  Epsilonproteobacteria  Campylobacterota  Campylobacter  \n",
       "1    Gammaproteobacteria    Pseudomonadota    Pseudomonas  \n",
       "2    Gammaproteobacteria    Pseudomonadota    Xanthomonas  \n",
       "3    Gammaproteobacteria    Pseudomonadota     Salmonella  \n",
       "4    Gammaproteobacteria    Pseudomonadota       Yersinia  \n",
       "\n",
       "[5 rows x 35 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prokaryotes_genomes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d0156271",
   "metadata": {},
   "outputs": [],
   "source": [
    "prokaryotes_genomes['ProteomeFile'] = prokaryotes_genomes['GenomeFile'].apply(lambda x: x.replace('_genomic.fna', '_cds_proteins.faa'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "3be4e97b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the directories\n",
    "proteome_dir = '/Users/akshayonly/Work/Sequence-Data/Proteomes'\n",
    "\n",
    "# Get the list of all files in the directories\n",
    "all_proteome_files = set(os.listdir(proteome_dir))\n",
    "selected_proteome_files = set(prokaryotes_genomes['ProteomeFile'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "e1bb8c4e-f0ac-4903-a4f9-118f2c87cbd8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 42593 entries, 0 to 42592\n",
      "Data columns (total 36 columns):\n",
      " #   Column                Non-Null Count  Dtype  \n",
      "---  ------                --------------  -----  \n",
      " 0   #Organism/Name        42593 non-null  object \n",
      " 1   TaxID                 42593 non-null  int64  \n",
      " 2   BioProject Accession  42593 non-null  object \n",
      " 3   BioProject ID         42593 non-null  int64  \n",
      " 4   Group                 42593 non-null  object \n",
      " 5   SubGroup              42593 non-null  object \n",
      " 6   Size (Mb)             42593 non-null  float64\n",
      " 7   GC%                   42593 non-null  object \n",
      " 8   Replicons             42593 non-null  object \n",
      " 9   WGS                   42593 non-null  object \n",
      " 10  Scaffolds             42593 non-null  object \n",
      " 11  Genes                 42593 non-null  object \n",
      " 12  Proteins              42593 non-null  object \n",
      " 13  Release Date          42593 non-null  object \n",
      " 14  Modify Date           42593 non-null  object \n",
      " 15  Status                42593 non-null  object \n",
      " 16  Center                42549 non-null  object \n",
      " 17  BioSample Accession   42593 non-null  object \n",
      " 18  Assembly Accession    42593 non-null  object \n",
      " 19  Reference             42593 non-null  object \n",
      " 20  FTP Path              42593 non-null  object \n",
      " 21  Pubmed ID             42593 non-null  object \n",
      " 22  Strain                42593 non-null  object \n",
      " 23  GenomePath            42593 non-null  object \n",
      " 24  CDSPath               42593 non-null  object \n",
      " 25  GenomeFile            42593 non-null  object \n",
      " 26  CDSFile               42593 non-null  object \n",
      " 27  Organism              42593 non-null  object \n",
      " 28  Lineage               42593 non-null  object \n",
      " 29  LineageTaxIDs         42593 non-null  int64  \n",
      " 30  Clade                 17052 non-null  object \n",
      " 31  Order                 42574 non-null  object \n",
      " 32  Class                 42049 non-null  object \n",
      " 33  Phylum                42593 non-null  object \n",
      " 34  Genus                 42593 non-null  object \n",
      " 35  ProteomeFile          42593 non-null  object \n",
      "dtypes: float64(1), int64(3), object(32)\n",
      "memory usage: 11.7+ MB\n"
     ]
    }
   ],
   "source": [
    "prokaryotes_genomes.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a077dcd4-44f9-40ef-9d0f-68bf32c439de",
   "metadata": {},
   "outputs": [],
   "source": [
    "remaining_genomes = prokaryotes_genomes[~prokaryotes_genomes['ProteomeFile'].isin(all_proteome_files)]['GenomePath'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "1f59756e-2418-4144-9429-e3a94a76318e",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(os.path.join('/Users/akshayonly/Work', 'remaining_genomes_ftp.txt'), 'w') as f:\n",
    "    for ftp in remaining_genomes:\n",
    "        f.write(f\"{ftp}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a8fe774-3d59-4056-a82d-c2caac1cd2cb",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
