{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from tqdm import tqdm\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_information_from_header(header, patterns):\n",
    "    \"\"\"Extracts information from a header string based on provided regex patterns.\"\"\"\n",
    "    return {key: re.search(pattern, header).group(1) if re.search(pattern, header) else None for key, pattern in patterns.items()}\n",
    "\n",
    "def format_gene_symbol(gene_symbol):\n",
    "    \"\"\"Formats the gene symbol by removing specified characters and adjusting casing.\"\"\"\n",
    "    return gene_symbol.lower().replace('[h','').replace('[c','').strip().replace('nuo', '').upper().translate(str.maketrans('', '', '-_/'))\n",
    "\n",
    "def parse_fasta_files(cds_dir, patterns):\n",
    "    \"\"\"Parses fasta files, extracting records with 'gene=nuo' and returns a DataFrame with extracted information.\"\"\"\n",
    "    data = []\n",
    "\n",
    "    for fasta in tqdm(os.listdir(cds_dir), desc=\"Parsing FASTA files\"):\n",
    "        for record in SeqIO.parse(os.path.join(cds_dir, fasta), \"fasta\"):\n",
    "            if \"gene=nuo\" in record.description:\n",
    "                prot_seq = record.seq.translate(table=11, to_stop=True)\n",
    "                data.append([fasta, record.description] + list(extract_information_from_header(record.description, patterns).values()) + [len(prot_seq)])\n",
    "    return pd.DataFrame(data, columns=['CDSFile', 'Header', 'Accession', 'GeneName', 'ProteinName', 'ProteinLength'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Parsing FASTA files: 100%|â–ˆ| 39893/39893 [08:24<00:00, 79.0\n"
     ]
    }
   ],
   "source": [
    "cds_dir = \"/Users/akshayonly/Work/Sequence-Data/CDS\"\n",
    "patterns = {\n",
    "    \"accession_number\": r\"lcl\\|(.*?)_cds\",\n",
    "    \"gene\": r\"\\[gene=(.*?)\\]\",\n",
    "    \"protein\": r\"\\[protein=(.*?)\\]\"\n",
    "}\n",
    "\n",
    "nuo_cds = parse_fasta_files(cds_dir, patterns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "nuo_cds['GeneName'] = nuo_cds['GeneName'].fillna('NONE').apply(format_gene_symbol)\n",
    "nuo_cds.drop_duplicates(subset=['CDSFile', 'GeneName'], inplace=True)\n",
    "\n",
    "# Preprocessing steps simplified and vectorized\n",
    "nuo_cds['Subunit'] = \"Nuo\" + nuo_cds['GeneName'].str.replace(r'\\d', '', regex=True)\n",
    "nuo_cds = nuo_cds[~nuo_cds['Subunit'].isin(['NuoBC', 'Nuo', 'NuoII', 'NuoLM', 'NuoP'])]\n",
    "\n",
    "# Define fused names and update 'Subunit'\n",
    "fused_names = \"|\".join([\"subunit C,D\", 'subunit C/D', 'subunit C/ D', 'chain C,D', 'chain C/D', 'chain C/ D', 'chain C; chain D', \n",
    "                        \"chain C, D\", \"chain CD\", \"subunit CD\", 'NADH-ubiquinone oxidoreductase chain C / NADH-ubiquinone oxidoreductase chain D', \n",
    "                        'NADH-quinone oxidoreductase, C/D subunit', 'NADH:ubiquinone oxidoreductase, fused CD subunit', 'NADH-ubiquinone oxidoreductase chain C'])\n",
    "\n",
    "nuo_cds.loc[nuo_cds['Subunit'].eq('NuoC') & nuo_cds['ProteinName'].str.contains(fused_names, case=False, na=False), 'Subunit'] = \"NuoCD\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "prok_data = pd.read_csv('/Users/akshayonly/Work/Updated/Data/01/prokaryotes_processed.csv')\n",
    "\n",
    "genome_data = pd.read_csv('/Users/akshayonly/Work/Updated/Data/01/genome_information.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "nuo_cds = pd.merge(nuo_cds, genome_data, on='Accession')\n",
    "\n",
    "nuo_cds = pd.merge(nuo_cds, prok_data, on='GenomeFile')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "nuo_cds = nuo_cds[['Lineage', 'LineageTaxIDs', 'Strain', 'Accession', 'Replicon', 'CDSFile', 'Header', 'ProteinLength', 'Subunit']]\n",
    "\n",
    "nuo_cds.drop_duplicates(inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Lineage</th>\n",
       "      <th>LineageTaxIDs</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Replicon</th>\n",
       "      <th>CDSFile</th>\n",
       "      <th>Header</th>\n",
       "      <th>ProteinLength</th>\n",
       "      <th>Subunit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>28901</td>\n",
       "      <td>CRSE-01</td>\n",
       "      <td>NZ_CP126166.1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GCF_030168445.1_ASM3016844v1_cds_from_genomic.fna</td>\n",
       "      <td>lcl|NZ_CP126166.1_cds_WP_000062993.1_2864 [gen...</td>\n",
       "      <td>147</td>\n",
       "      <td>NuoA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>28901</td>\n",
       "      <td>CRSE-01</td>\n",
       "      <td>NZ_CP126166.1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GCF_030168445.1_ASM3016844v1_cds_from_genomic.fna</td>\n",
       "      <td>lcl|NZ_CP126166.1_cds_WP_000386728.1_2865 [gen...</td>\n",
       "      <td>220</td>\n",
       "      <td>NuoB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>28901</td>\n",
       "      <td>CRSE-01</td>\n",
       "      <td>NZ_CP126166.1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GCF_030168445.1_ASM3016844v1_cds_from_genomic.fna</td>\n",
       "      <td>lcl|NZ_CP126166.1_cds_WP_000247855.1_2866 [gen...</td>\n",
       "      <td>600</td>\n",
       "      <td>NuoCD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>28901</td>\n",
       "      <td>CRSE-01</td>\n",
       "      <td>NZ_CP126166.1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GCF_030168445.1_ASM3016844v1_cds_from_genomic.fna</td>\n",
       "      <td>lcl|NZ_CP126166.1_cds_WP_000545038.1_2867 [gen...</td>\n",
       "      <td>166</td>\n",
       "      <td>NuoE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>28901</td>\n",
       "      <td>CRSE-01</td>\n",
       "      <td>NZ_CP126166.1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GCF_030168445.1_ASM3016844v1_cds_from_genomic.fna</td>\n",
       "      <td>lcl|NZ_CP126166.1_cds_WP_000800048.1_2868 [gen...</td>\n",
       "      <td>445</td>\n",
       "      <td>NuoF</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Lineage  LineageTaxIDs   Strain      Accession    Replicon  \\\n",
       "0  Salmonella enterica          28901  CRSE-01  NZ_CP126166.1  Chromosome   \n",
       "1  Salmonella enterica          28901  CRSE-01  NZ_CP126166.1  Chromosome   \n",
       "2  Salmonella enterica          28901  CRSE-01  NZ_CP126166.1  Chromosome   \n",
       "3  Salmonella enterica          28901  CRSE-01  NZ_CP126166.1  Chromosome   \n",
       "4  Salmonella enterica          28901  CRSE-01  NZ_CP126166.1  Chromosome   \n",
       "\n",
       "                                             CDSFile  \\\n",
       "0  GCF_030168445.1_ASM3016844v1_cds_from_genomic.fna   \n",
       "1  GCF_030168445.1_ASM3016844v1_cds_from_genomic.fna   \n",
       "2  GCF_030168445.1_ASM3016844v1_cds_from_genomic.fna   \n",
       "3  GCF_030168445.1_ASM3016844v1_cds_from_genomic.fna   \n",
       "4  GCF_030168445.1_ASM3016844v1_cds_from_genomic.fna   \n",
       "\n",
       "                                              Header  ProteinLength Subunit  \n",
       "0  lcl|NZ_CP126166.1_cds_WP_000062993.1_2864 [gen...            147    NuoA  \n",
       "1  lcl|NZ_CP126166.1_cds_WP_000386728.1_2865 [gen...            220    NuoB  \n",
       "2  lcl|NZ_CP126166.1_cds_WP_000247855.1_2866 [gen...            600   NuoCD  \n",
       "3  lcl|NZ_CP126166.1_cds_WP_000545038.1_2867 [gen...            166    NuoE  \n",
       "4  lcl|NZ_CP126166.1_cds_WP_000800048.1_2868 [gen...            445    NuoF  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nuo_cds.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "uniprot = pd.read_csv('/Users/akshayonly/Work/Updated/Data/Misc/Uniprot/uniprot_reviewed_nuo_proteins_length.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge this information with cds_chromosome\n",
    "merged_data = nuo_cds.merge(uniprot[['Subunit', 'MinLength']], on='Subunit', how='left')\n",
    "\n",
    "# Use a vectorized operation to assign \"To-Review\" or \"Reviewed\" based on the condition\n",
    "merged_data['ReviewStatus'] = \"To-Review\"  # Default value for all rows\n",
    "# Update only the rows that meet the condition\n",
    "merged_data.loc[merged_data['ProteinLength'] >= merged_data['MinLength'], 'ReviewStatus'] = \"Reviewed\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Lineage</th>\n",
       "      <th>LineageTaxIDs</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Replicon</th>\n",
       "      <th>CDSFile</th>\n",
       "      <th>Header</th>\n",
       "      <th>ProteinLength</th>\n",
       "      <th>Subunit</th>\n",
       "      <th>MinLength</th>\n",
       "      <th>ReviewStatus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>28901</td>\n",
       "      <td>CRSE-01</td>\n",
       "      <td>NZ_CP126166.1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GCF_030168445.1_ASM3016844v1_cds_from_genomic.fna</td>\n",
       "      <td>lcl|NZ_CP126166.1_cds_WP_000062993.1_2864 [gen...</td>\n",
       "      <td>147</td>\n",
       "      <td>NuoA</td>\n",
       "      <td>116.0</td>\n",
       "      <td>Reviewed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>28901</td>\n",
       "      <td>CRSE-01</td>\n",
       "      <td>NZ_CP126166.1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GCF_030168445.1_ASM3016844v1_cds_from_genomic.fna</td>\n",
       "      <td>lcl|NZ_CP126166.1_cds_WP_000386728.1_2865 [gen...</td>\n",
       "      <td>220</td>\n",
       "      <td>NuoB</td>\n",
       "      <td>158.0</td>\n",
       "      <td>Reviewed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>28901</td>\n",
       "      <td>CRSE-01</td>\n",
       "      <td>NZ_CP126166.1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GCF_030168445.1_ASM3016844v1_cds_from_genomic.fna</td>\n",
       "      <td>lcl|NZ_CP126166.1_cds_WP_000247855.1_2866 [gen...</td>\n",
       "      <td>600</td>\n",
       "      <td>NuoCD</td>\n",
       "      <td>519.0</td>\n",
       "      <td>Reviewed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>28901</td>\n",
       "      <td>CRSE-01</td>\n",
       "      <td>NZ_CP126166.1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GCF_030168445.1_ASM3016844v1_cds_from_genomic.fna</td>\n",
       "      <td>lcl|NZ_CP126166.1_cds_WP_000545038.1_2867 [gen...</td>\n",
       "      <td>166</td>\n",
       "      <td>NuoE</td>\n",
       "      <td>160.0</td>\n",
       "      <td>Reviewed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>28901</td>\n",
       "      <td>CRSE-01</td>\n",
       "      <td>NZ_CP126166.1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GCF_030168445.1_ASM3016844v1_cds_from_genomic.fna</td>\n",
       "      <td>lcl|NZ_CP126166.1_cds_WP_000800048.1_2868 [gen...</td>\n",
       "      <td>445</td>\n",
       "      <td>NuoF</td>\n",
       "      <td>417.0</td>\n",
       "      <td>Reviewed</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Lineage  LineageTaxIDs   Strain      Accession    Replicon  \\\n",
       "0  Salmonella enterica          28901  CRSE-01  NZ_CP126166.1  Chromosome   \n",
       "1  Salmonella enterica          28901  CRSE-01  NZ_CP126166.1  Chromosome   \n",
       "2  Salmonella enterica          28901  CRSE-01  NZ_CP126166.1  Chromosome   \n",
       "3  Salmonella enterica          28901  CRSE-01  NZ_CP126166.1  Chromosome   \n",
       "4  Salmonella enterica          28901  CRSE-01  NZ_CP126166.1  Chromosome   \n",
       "\n",
       "                                             CDSFile  \\\n",
       "0  GCF_030168445.1_ASM3016844v1_cds_from_genomic.fna   \n",
       "1  GCF_030168445.1_ASM3016844v1_cds_from_genomic.fna   \n",
       "2  GCF_030168445.1_ASM3016844v1_cds_from_genomic.fna   \n",
       "3  GCF_030168445.1_ASM3016844v1_cds_from_genomic.fna   \n",
       "4  GCF_030168445.1_ASM3016844v1_cds_from_genomic.fna   \n",
       "\n",
       "                                              Header  ProteinLength Subunit  \\\n",
       "0  lcl|NZ_CP126166.1_cds_WP_000062993.1_2864 [gen...            147    NuoA   \n",
       "1  lcl|NZ_CP126166.1_cds_WP_000386728.1_2865 [gen...            220    NuoB   \n",
       "2  lcl|NZ_CP126166.1_cds_WP_000247855.1_2866 [gen...            600   NuoCD   \n",
       "3  lcl|NZ_CP126166.1_cds_WP_000545038.1_2867 [gen...            166    NuoE   \n",
       "4  lcl|NZ_CP126166.1_cds_WP_000800048.1_2868 [gen...            445    NuoF   \n",
       "\n",
       "   MinLength ReviewStatus  \n",
       "0      116.0     Reviewed  \n",
       "1      158.0     Reviewed  \n",
       "2      519.0     Reviewed  \n",
       "3      160.0     Reviewed  \n",
       "4      417.0     Reviewed  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "nuo_cds = merged_data.drop(columns=['MinLength'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "nuo_cds.to_csv('/Users/akshayonly/Work/Updated/Data/02/prescreened_nuo_entries_from_cds.csv', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
