{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "12acf173-10e2-4c95-b7d8-8fc3157434a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from tqdm import tqdm\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9ef4f279-5119-4b76-a079-9680b037abb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.read_csv('/Users/akshayonly/Work/Updated/Data/02/prescreened_nuo_entries_from_cds.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a7350666-4a38-47fc-8d11-3f854ea732ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Regex pattern to extract numbers\n",
    "pattern = re.compile(r'\\d+')\n",
    "\n",
    "# Extract numbers from each string\n",
    "extracted_numbers = [pattern.findall(item) for item in data['Header'].apply(lambda x: x.split(' ')[-2]).values]\n",
    "\n",
    "extracted_numbers = pd.DataFrame(extracted_numbers).iloc[:, :-2]\n",
    "\n",
    "extracted_numbers.columns = ['Start', 'End']\n",
    "\n",
    "data = pd.concat([data, extracted_numbers], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "367a4aff-edc7-4745-a6c4-a290fab8d74b",
   "metadata": {},
   "outputs": [],
   "source": [
    "data['Start'] = data['Start'].astype(int)\n",
    "data['End'] = data['End'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e635ae73-acb7-4893-a493-148d2a356430",
   "metadata": {},
   "outputs": [],
   "source": [
    "def cluster_genes(group, threshold=250):\n",
    "    # Sort the group by 'Start' position\n",
    "    group = group.sort_values('Start')\n",
    "    \n",
    "    # Initialize cluster labels\n",
    "    group['Cluster'] = 0\n",
    "    \n",
    "    # Calculate distances between consecutive genes\n",
    "    distances = group['Start'].values[1:] - group['End'].values[:-1]\n",
    "    \n",
    "    # Determine cluster IDs (increment when distance >= threshold)\n",
    "    group['Cluster'][1:] = (distances >= threshold).cumsum()\n",
    "    \n",
    "    return group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1aeb6027-1379-4547-b4b1-9edd328dbf09",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_cls = data.groupby('Accession').apply(cluster_genes).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "039e8c8a-2955-4274-b4c1-c747d5573bda",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_subunits = sorted(data_cls['Subunit'].unique())\n",
    "\n",
    "# Create a pivot table with 'Accession' and 'Cluster' as indices\n",
    "pivot = data_cls.pivot_table(index=['Accession', 'Cluster'], \n",
    "                          columns='Subunit', \n",
    "                          aggfunc='size', \n",
    "                          fill_value=0)\n",
    "\n",
    "# Ensure all subunits are present as columns, even if no data exists for them\n",
    "pivot = pivot.reindex(columns=all_subunits, fill_value=0).reset_index()\n",
    "\n",
    "# Rename the columns as per your requirement (if necessary)\n",
    "nuo_count = pd.DataFrame(pivot)\n",
    "nuo_count.columns = ['Accession', 'Cluster'] + all_subunits  # Adjust if specific names are needed\n",
    "\n",
    "# Create a copy of the dataframe to apply boolean conversion\n",
    "nuo_bool = nuo_count.copy()\n",
    "nuo_bool[nuo_bool.iloc[:, 2:].columns] = nuo_bool.iloc[:, 2:].ge(1)\n",
    "\n",
    "nuo_bool['Count'] = nuo_bool[all_subunits].sum(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ca09f240-2eca-439c-ad18-0a2d8f74b56e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Lineage</th>\n",
       "      <th>LineageTaxIDs</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Replicon</th>\n",
       "      <th>CDSFile</th>\n",
       "      <th>Header</th>\n",
       "      <th>ProteinLength</th>\n",
       "      <th>Subunit</th>\n",
       "      <th>ReviewStatus</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>562</td>\n",
       "      <td>2020CK-00188</td>\n",
       "      <td>ABACVG020000001.1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GCA_018071945.2_ASM1807194v2_cds_from_genomic.fna</td>\n",
       "      <td>lcl|ABACVG020000001.1_cds_MCZ6945865.1_824 [ge...</td>\n",
       "      <td>485</td>\n",
       "      <td>NuoN</td>\n",
       "      <td>Reviewed</td>\n",
       "      <td>876770</td>\n",
       "      <td>878227</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>562</td>\n",
       "      <td>2020CK-00188</td>\n",
       "      <td>ABACVG020000001.1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GCA_018071945.2_ASM1807194v2_cds_from_genomic.fna</td>\n",
       "      <td>lcl|ABACVG020000001.1_cds_MCZ6945866.1_825 [ge...</td>\n",
       "      <td>509</td>\n",
       "      <td>NuoM</td>\n",
       "      <td>Reviewed</td>\n",
       "      <td>878234</td>\n",
       "      <td>879763</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>562</td>\n",
       "      <td>2020CK-00188</td>\n",
       "      <td>ABACVG020000001.1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GCA_018071945.2_ASM1807194v2_cds_from_genomic.fna</td>\n",
       "      <td>lcl|ABACVG020000001.1_cds_MCZ6945867.1_826 [ge...</td>\n",
       "      <td>613</td>\n",
       "      <td>NuoL</td>\n",
       "      <td>Reviewed</td>\n",
       "      <td>879994</td>\n",
       "      <td>881835</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>562</td>\n",
       "      <td>2020CK-00188</td>\n",
       "      <td>ABACVG020000001.1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GCA_018071945.2_ASM1807194v2_cds_from_genomic.fna</td>\n",
       "      <td>lcl|ABACVG020000001.1_cds_MCZ6945868.1_827 [ge...</td>\n",
       "      <td>100</td>\n",
       "      <td>NuoK</td>\n",
       "      <td>Reviewed</td>\n",
       "      <td>881832</td>\n",
       "      <td>882134</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>562</td>\n",
       "      <td>2020CK-00188</td>\n",
       "      <td>ABACVG020000001.1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GCA_018071945.2_ASM1807194v2_cds_from_genomic.fna</td>\n",
       "      <td>lcl|ABACVG020000001.1_cds_MCZ6945869.1_828 [ge...</td>\n",
       "      <td>184</td>\n",
       "      <td>NuoJ</td>\n",
       "      <td>Reviewed</td>\n",
       "      <td>882131</td>\n",
       "      <td>882685</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Lineage  LineageTaxIDs        Strain          Accession  \\\n",
       "0  Escherichia coli            562  2020CK-00188  ABACVG020000001.1   \n",
       "1  Escherichia coli            562  2020CK-00188  ABACVG020000001.1   \n",
       "2  Escherichia coli            562  2020CK-00188  ABACVG020000001.1   \n",
       "3  Escherichia coli            562  2020CK-00188  ABACVG020000001.1   \n",
       "4  Escherichia coli            562  2020CK-00188  ABACVG020000001.1   \n",
       "\n",
       "     Replicon                                            CDSFile  \\\n",
       "0  Chromosome  GCA_018071945.2_ASM1807194v2_cds_from_genomic.fna   \n",
       "1  Chromosome  GCA_018071945.2_ASM1807194v2_cds_from_genomic.fna   \n",
       "2  Chromosome  GCA_018071945.2_ASM1807194v2_cds_from_genomic.fna   \n",
       "3  Chromosome  GCA_018071945.2_ASM1807194v2_cds_from_genomic.fna   \n",
       "4  Chromosome  GCA_018071945.2_ASM1807194v2_cds_from_genomic.fna   \n",
       "\n",
       "                                              Header  ProteinLength Subunit  \\\n",
       "0  lcl|ABACVG020000001.1_cds_MCZ6945865.1_824 [ge...            485    NuoN   \n",
       "1  lcl|ABACVG020000001.1_cds_MCZ6945866.1_825 [ge...            509    NuoM   \n",
       "2  lcl|ABACVG020000001.1_cds_MCZ6945867.1_826 [ge...            613    NuoL   \n",
       "3  lcl|ABACVG020000001.1_cds_MCZ6945868.1_827 [ge...            100    NuoK   \n",
       "4  lcl|ABACVG020000001.1_cds_MCZ6945869.1_828 [ge...            184    NuoJ   \n",
       "\n",
       "  ReviewStatus   Start     End  Cluster  \n",
       "0     Reviewed  876770  878227        0  \n",
       "1     Reviewed  878234  879763        0  \n",
       "2     Reviewed  879994  881835        0  \n",
       "3     Reviewed  881832  882134        0  \n",
       "4     Reviewed  882131  882685        0  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_cls.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e758f68d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Lineage</th>\n",
       "      <th>LineageTaxIDs</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Replicon</th>\n",
       "      <th>CDSFile</th>\n",
       "      <th>Header</th>\n",
       "      <th>ProteinLength</th>\n",
       "      <th>Subunit</th>\n",
       "      <th>ReviewStatus</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>28901</td>\n",
       "      <td>CRSE-01</td>\n",
       "      <td>NZ_CP126166.1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GCF_030168445.1_ASM3016844v1_cds_from_genomic.fna</td>\n",
       "      <td>lcl|NZ_CP126166.1_cds_WP_000062993.1_2864 [gen...</td>\n",
       "      <td>147</td>\n",
       "      <td>NuoA</td>\n",
       "      <td>Reviewed</td>\n",
       "      <td>3108466</td>\n",
       "      <td>3108909</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Lineage  LineageTaxIDs   Strain      Accession    Replicon  \\\n",
       "0  Salmonella enterica          28901  CRSE-01  NZ_CP126166.1  Chromosome   \n",
       "\n",
       "                                             CDSFile  \\\n",
       "0  GCF_030168445.1_ASM3016844v1_cds_from_genomic.fna   \n",
       "\n",
       "                                              Header  ProteinLength Subunit  \\\n",
       "0  lcl|NZ_CP126166.1_cds_WP_000062993.1_2864 [gen...            147    NuoA   \n",
       "\n",
       "  ReviewStatus    Start      End  \n",
       "0     Reviewed  3108466  3108909  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a732c109",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Global configurations\n",
    "DATA_DIR = \"/Users/akshayonly/Work/Sequence-Data/CDS\"\n",
    "OUTPUT_DIR = \"/Users/akshayonly/Work/Sequence-Data/Nuo-Sequences\"\n",
    "\n",
    "if not os.path.exists(OUTPUT_DIR):\n",
    "    os.makedirs(OUTPUT_DIR)\n",
    "\n",
    "# Function to filter and process sequence data for an organism\n",
    "def process_sequences(cds, cds_dir):\n",
    "    \"\"\"\n",
    "    Process and save sequences for a given organism.\n",
    "    :param organism: Organism type ('Bacteria' or 'Archaea')\n",
    "    :param cds: DataFrame containing CDS information\n",
    "    :param cds_dir: Directory containing CDS files\n",
    "    \"\"\"\n",
    "    # Filter CDS entries for the organism\n",
    "    chr_cds = cds[(cds['Replicon']=='Chromosome') & (cds['ReviewStatus']=='Reviewed')]\n",
    "    cds_to_info = chr_cds.groupby('CDSFile')[['Subunit', 'Header']].apply(lambda x: x.values.tolist()).to_dict()\n",
    "    sequence_data = {subunit: [] for subunit in cds['Subunit'].unique()}\n",
    "\n",
    "    # Process each CDS file\n",
    "    for fasta in tqdm(chr_cds['CDSFile'].unique()):\n",
    "        fasta_fullpath = os.path.join(cds_dir, fasta)\n",
    "        for record in SeqIO.parse(fasta_fullpath, \"fasta\"):\n",
    "            for subunit, header in cds_to_info[fasta]:\n",
    "                if header in record.description:\n",
    "                    sequence_data[subunit].append(record.seq.translate(table=11, to_stop=True))\n",
    "\n",
    "    # Write sequences to files\n",
    "    for subunit, sequences in sequence_data.items():\n",
    "        records = [SeqRecord(seq, id=f\"{subunit}_{i+1}\", description=f\"{subunit} Subunit\") for i, seq in enumerate(sequences)]\n",
    "        if records:\n",
    "            SeqIO.write(records, os.path.join(OUTPUT_DIR, f\"{subunit.lower()}_cds.faa\"), \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2ea72020",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 24696/24696 [06:56<00:00, 59.26it/s]\n"
     ]
    }
   ],
   "source": [
    "process_sequences(data, DATA_DIR)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
