{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import statistics as st\n",
    "\n",
    "from tqdm import tqdm\n",
    "from glob import glob\n",
    "\n",
    "import re\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "import warnings\n",
    "\n",
    "warnings.filterwarnings('ignore')  # Ignore all warnings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cluster_genes(group, threshold=250):\n",
    "    # Sort the group by 'Start' position\n",
    "    group = group.sort_values('Start')\n",
    "    \n",
    "    # Initialize cluster labels\n",
    "    group['Cluster'] = 0\n",
    "    \n",
    "    # Calculate distances between consecutive genes\n",
    "    distances = group['Start'].values[1:] - group['End'].values[:-1]\n",
    "    \n",
    "    # Determine cluster IDs (increment when distance >= threshold)\n",
    "    group['Cluster'][1:] = (distances >= threshold).cumsum()\n",
    "    \n",
    "    return group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_subunit_data(hmmer):\n",
    "    # Generate a list of all unique 'Subunit' values, sorted\n",
    "    all_subunits = ['NuoA', 'NuoB', 'NuoBCD', 'NuoC', \n",
    "                    'NuoCD', 'NuoD', 'NuoE', 'NuoF', \n",
    "                    'NuoG', 'NuoH', 'NuoI', 'NuoJ', \n",
    "                    'NuoK', 'NuoL', 'NuoM', 'NuoN']\n",
    "\n",
    "    # Create a pivot table with 'Accession' and 'Cluster' as indices\n",
    "    pivot = hmmer.pivot_table(index=['Accession', 'Cluster'], \n",
    "                              columns='Subunit', \n",
    "                              aggfunc='size', \n",
    "                              fill_value=0)\n",
    "\n",
    "    # Ensure all subunits are present as columns, even if no data exists for them\n",
    "    pivot = pivot.reindex(columns=all_subunits, fill_value=0).reset_index()\n",
    "\n",
    "    # Rename the columns as per your requirement (if necessary)\n",
    "    nuo_count = pd.DataFrame(pivot)\n",
    "    nuo_count.columns = ['Accession', 'Cluster'] + all_subunits  # Adjust if specific names are needed\n",
    "\n",
    "    # Create a copy of the dataframe to apply boolean conversion\n",
    "    nuo_bool = nuo_count.copy()\n",
    "    nuo_bool[nuo_bool.iloc[:, 2:].columns] = nuo_bool.iloc[:, 2:].ge(1)\n",
    "\n",
    "    nuo_bool['Count'] = nuo_bool[all_subunits].sum(axis=1)\n",
    "\n",
    "    return nuo_count, nuo_bool"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "prok_info = pd.read_csv('/Users/akshayonly/Work/Updated/Data/01/prokaryotes_processed.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "hmmer = pd.read_csv('/Users/akshayonly/Work/Updated/Data/hmmer_search_results_processed.csv', low_memory=False)\n",
    "metadata = pd.read_csv('/Users/akshayonly/Work/Updated/Data/01/genome_information.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = pd.merge(prok_info[['GenomeFile', 'Organism', 'Lineage', 'LineageTaxIDs', 'Strain', 'TaxID']], metadata, on='GenomeFile')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata.drop_duplicates(inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Apply clustering function to each subset of data grouped by 'Accession'\n",
    "hmmer = hmmer.groupby('Accession').apply(cluster_genes).reset_index(drop=True)\n",
    "\n",
    "hmmer = pd.merge(metadata[['Accession', 'ProteomeFile']], hmmer, on='Accession')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 756835 entries, 0 to 756834\n",
      "Data columns (total 11 columns):\n",
      " #   Column            Non-Null Count   Dtype  \n",
      "---  ------            --------------   -----  \n",
      " 0   Accession         756835 non-null  object \n",
      " 1   ProteomeFile      756835 non-null  object \n",
      " 2   ProteinAccession  756835 non-null  object \n",
      " 3   Subunit           756835 non-null  object \n",
      " 4   evalue            756835 non-null  float64\n",
      " 5   BitScore          756835 non-null  float64\n",
      " 6   Bias              756835 non-null  float64\n",
      " 7   SequenceDesc      756835 non-null  object \n",
      " 8   Start             756835 non-null  int64  \n",
      " 9   End               756835 non-null  int64  \n",
      " 10  Cluster           756835 non-null  int64  \n",
      "dtypes: float64(3), int64(3), object(5)\n",
      "memory usage: 63.5+ MB\n"
     ]
    }
   ],
   "source": [
    "hmmer.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GenomeFile</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Lineage</th>\n",
       "      <th>LineageTaxIDs</th>\n",
       "      <th>Strain</th>\n",
       "      <th>TaxID</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Replicon</th>\n",
       "      <th>ProteomeFile</th>\n",
       "      <th>SequenceLength(Mb)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_000009085.1_ASM908v1_genomic.fna</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Campylobacter jejuni</td>\n",
       "      <td>197</td>\n",
       "      <td>NCTC 11168</td>\n",
       "      <td>192222</td>\n",
       "      <td>AL111168.1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GCA_000009085.1_ASM908v1_cds_proteins.faa</td>\n",
       "      <td>1.641481</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             GenomeFile  Organism               Lineage  \\\n",
       "0  GCA_000009085.1_ASM908v1_genomic.fna  Bacteria  Campylobacter jejuni   \n",
       "\n",
       "   LineageTaxIDs      Strain   TaxID   Accession    Replicon  \\\n",
       "0            197  NCTC 11168  192222  AL111168.1  Chromosome   \n",
       "\n",
       "                                ProteomeFile  SequenceLength(Mb)  \n",
       "0  GCA_000009085.1_ASM908v1_cds_proteins.faa            1.641481  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "nuo_count, nuo_bool = generate_subunit_data(hmmer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "41277"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hmmer['Accession'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "41277"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nuo_bool['Accession'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "rearrange_columns = ['Accession', 'Cluster', 'Count', 'NuoA', 'NuoB', 'NuoBCD', \n",
    "                     'NuoC', 'NuoCD', 'NuoD', 'NuoE', 'NuoF', 'NuoG', 'NuoH', 'NuoI', 'NuoJ', \n",
    "                     'NuoK', 'NuoL', 'NuoM', 'NuoN']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "nuo_bool = nuo_bool[rearrange_columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|███████████████████████| 2/2 [00:00<00:00, 146.22it/s]\n"
     ]
    }
   ],
   "source": [
    "# Defining the combinations\n",
    "combinations = {\n",
    "    'Complete 14 Subunits': ['NuoA', 'NuoB', 'NuoC', 'NuoD', 'NuoE', 'NuoF', 'NuoG', 'NuoH', 'NuoI', 'NuoJ', 'NuoK', 'NuoL', 'NuoM', 'NuoN'],\n",
    "    '13 Subunits with Fused CD': ['NuoA', 'NuoB', 'NuoCD', 'NuoE', 'NuoF', 'NuoG', 'NuoH', 'NuoI', 'NuoJ', 'NuoK', 'NuoL', 'NuoM', 'NuoN'],\n",
    "    # '12 Subunits with Fused BCD': ['NuoA', 'NuoBCD', 'NuoE', 'NuoF', 'NuoG', 'NuoH', 'NuoI', 'NuoJ', 'NuoK', 'NuoL', 'NuoM', 'NuoN'],\n",
    "    # '12 Subunits Excluding NuoEF': ['NuoA', 'NuoB', 'NuoC', 'NuoD', 'NuoG', 'NuoH', 'NuoI', 'NuoJ', 'NuoK', 'NuoL', 'NuoM', 'NuoN'],\n",
    "    # '11 Subunits with Fused CD Excluding NuoEF': ['NuoA', 'NuoB', 'NuoCD', 'NuoG', 'NuoH', 'NuoI', 'NuoJ', 'NuoK', 'NuoL', 'NuoM', 'NuoN'],\n",
    "    # '11 Subunits Excluding NuoEFG': ['NuoA', 'NuoB', 'NuoC', 'NuoD', 'NuoH', 'NuoI', 'NuoJ', 'NuoK', 'NuoL', 'NuoM', 'NuoN'],\n",
    "    # '11 Subunits with Fused CD Excluding NuoEFG': ['NuoA', 'NuoB', 'NuoCD', 'NuoH', 'NuoI', 'NuoJ', 'NuoK', 'NuoL', 'NuoM', 'NuoN']\n",
    "}\n",
    "\n",
    "all_subunits = nuo_bool.iloc[:, 3:].columns.tolist()\n",
    "\n",
    "nuo_bool['ComplexType'] = 'Less than 11 Subunits'\n",
    "\n",
    "# Checking for each variation\n",
    "results = {}\n",
    "# Iterating over combinations to update the new column\n",
    "for name, true_subunits in tqdm(combinations.items()):\n",
    "    false_subunits = [subunit for subunit in all_subunits if subunit not in true_subunits]\n",
    "    \n",
    "    # Identifying rows that match the current combination\n",
    "    match_mask = (nuo_bool[true_subunits].all(axis=1)) & (~nuo_bool[false_subunits].any(axis=1))\n",
    "    \n",
    "    # Updating the 'CombinationType' column for rows that match the combination\n",
    "    nuo_bool.loc[match_mask, 'ComplexType'] = name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "selected_columns = ['Accession', 'Replicon', 'ProteomeFile', 'Organism', 'Lineage', 'LineageTaxIDs', 'Strain', 'TaxID']\n",
    "\n",
    "nuo_bool_detailed = nuo_bool.copy()\n",
    "nuo_bool_detailed = pd.merge(metadata[selected_columns], nuo_bool_detailed, on='Accession')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_leftout = metadata[~metadata['ProteomeFile'].isin(nuo_bool_detailed['ProteomeFile'])][selected_columns]\n",
    "\n",
    "metadata_leftout = metadata_leftout.assign(**{col: False for col in all_subunits})\n",
    "metadata_leftout['ComplexType'] = 'Absent'\n",
    "metadata_leftout['Cluster'] = None\n",
    "\n",
    "rearrange_columns = ['Accession', 'Replicon', 'ProteomeFile', 'Organism', 'Lineage', 'LineageTaxIDs', 'Strain', 'TaxID',\n",
    "                     'ComplexType', 'Cluster', 'NuoA', 'NuoB', 'NuoBCD', 'NuoC', 'NuoCD', 'NuoD', 'NuoE', 'NuoF', \n",
    "                     'NuoG', 'NuoH', 'NuoI', 'NuoJ', 'NuoK', 'NuoL', 'NuoM', 'NuoN']\n",
    "\n",
    "nuo_bool_detailed = nuo_bool_detailed[rearrange_columns]\n",
    "metadata_leftout = metadata_leftout[rearrange_columns]\n",
    "\n",
    "nuo_bool_detailed = pd.concat([nuo_bool_detailed, metadata_leftout])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "incomplete_set = set(nuo_bool_detailed[(nuo_bool_detailed['ComplexType'].isin(['Less than 11 Subunits','Absent']))]['LineageTaxIDs'])\n",
    "\n",
    "complete_set = set(nuo_bool_detailed[(nuo_bool_detailed['ComplexType'].isin(['Complete 14 Subunits', '13 Subunits with Fused CD']))]['LineageTaxIDs'])\n",
    "\n",
    "# Create a boolean mask for rows to remove\n",
    "mask = (nuo_bool_detailed['ComplexType'].isin(['Less than 11 Subunits','Absent'])) & (nuo_bool_detailed['LineageTaxIDs'].isin(incomplete_set.intersection(complete_set)))\n",
    "\n",
    "# Use the negation of the mask to keep rows that do not meet the condition\n",
    "nuo_bool_detailed = nuo_bool_detailed[~mask]\n",
    "\n",
    "# Optionally reset the index if you want a clean DataFrame index\n",
    "nuo_bool_detailed.reset_index(drop=True, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13 Subunits with Fused CD 1345 12726\n",
      "Absent 801 5426\n",
      "Complete 14 Subunits 2610 7201\n",
      "Less than 11 Subunits 5160 15239\n"
     ]
    }
   ],
   "source": [
    "for complex_type in sorted(nuo_bool_detailed['ComplexType'].unique()):\n",
    "    no_lineages = nuo_bool_detailed[nuo_bool_detailed['ComplexType']==complex_type]['LineageTaxIDs'].nunique()\n",
    "    no_genomes = nuo_bool_detailed[nuo_bool_detailed['ComplexType']==complex_type]['ProteomeFile'].nunique()\n",
    "    print(complex_type, no_lineages, no_genomes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Replicon</th>\n",
       "      <th>ProteomeFile</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Lineage</th>\n",
       "      <th>LineageTaxIDs</th>\n",
       "      <th>Strain</th>\n",
       "      <th>TaxID</th>\n",
       "      <th>ComplexType</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>...</th>\n",
       "      <th>NuoE</th>\n",
       "      <th>NuoF</th>\n",
       "      <th>NuoG</th>\n",
       "      <th>NuoH</th>\n",
       "      <th>NuoI</th>\n",
       "      <th>NuoJ</th>\n",
       "      <th>NuoK</th>\n",
       "      <th>NuoL</th>\n",
       "      <th>NuoM</th>\n",
       "      <th>NuoN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>LT907842.1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GCA_900215245.1_IMG-taxon_2617270901_annotated...</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Pseudomonas fluorescens</td>\n",
       "      <td>294</td>\n",
       "      <td>ATCC 13525</td>\n",
       "      <td>294</td>\n",
       "      <td>13 Subunits with Fused CD</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CP058243.1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GCA_013388375.1_ASM1338837v1_cds_proteins.faa</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Xanthomonas campestris</td>\n",
       "      <td>339</td>\n",
       "      <td>MAFF106181</td>\n",
       "      <td>359385</td>\n",
       "      <td>Complete 14 Subunits</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CP002956.1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GCA_000222975.1_ASM22297v1_cds_proteins.faa</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Yersinia pestis</td>\n",
       "      <td>632</td>\n",
       "      <td>A1122</td>\n",
       "      <td>1035377</td>\n",
       "      <td>13 Subunits with Fused CD</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>CP018019.1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GCA_022175585.1_ASM2217558v1_cds_proteins.faa</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Mycobacterium avium</td>\n",
       "      <td>1764</td>\n",
       "      <td>OCU889s_P11_4s</td>\n",
       "      <td>439334</td>\n",
       "      <td>Complete 14 Subunits</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>NZ_CP039850.1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GCF_005156105.1_ASM515610v1_cds_proteins.faa</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>1773</td>\n",
       "      <td>Danish 1331</td>\n",
       "      <td>1765</td>\n",
       "      <td>Complete 14 Subunits</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114499</th>\n",
       "      <td>CP133106.1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GCA_031236885.1_ASM3123688v1_cds_proteins.faa</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>562</td>\n",
       "      <td>8523</td>\n",
       "      <td>562</td>\n",
       "      <td>13 Subunits with Fused CD</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114500</th>\n",
       "      <td>CP085372.1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GCA_020673295.1_ASM2067329v1_cds_proteins.faa</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>562</td>\n",
       "      <td>NDM7</td>\n",
       "      <td>562</td>\n",
       "      <td>13 Subunits with Fused CD</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114501</th>\n",
       "      <td>CP092707.1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GCA_022488325.1_ASM2248832v1_cds_proteins.faa</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>562</td>\n",
       "      <td>S-P-N-065.01</td>\n",
       "      <td>562</td>\n",
       "      <td>13 Subunits with Fused CD</td>\n",
       "      <td>8</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114502</th>\n",
       "      <td>CP092704.1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GCA_022488265.1_ASM2248826v1_cds_proteins.faa</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>562</td>\n",
       "      <td>S-P-N-045.01</td>\n",
       "      <td>562</td>\n",
       "      <td>13 Subunits with Fused CD</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114503</th>\n",
       "      <td>CP092699.1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GCA_022488345.1_ASM2248834v1_cds_proteins.faa</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>562</td>\n",
       "      <td>S-P-N-063.01</td>\n",
       "      <td>562</td>\n",
       "      <td>13 Subunits with Fused CD</td>\n",
       "      <td>8</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19961 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Accession    Replicon  \\\n",
       "5          LT907842.1  Chromosome   \n",
       "6          CP058243.1  Chromosome   \n",
       "7          CP002956.1  Chromosome   \n",
       "31         CP018019.1  Chromosome   \n",
       "32      NZ_CP039850.1  Chromosome   \n",
       "...               ...         ...   \n",
       "114499     CP133106.1  Chromosome   \n",
       "114500     CP085372.1  Chromosome   \n",
       "114501     CP092707.1  Chromosome   \n",
       "114502     CP092704.1  Chromosome   \n",
       "114503     CP092699.1  Chromosome   \n",
       "\n",
       "                                             ProteomeFile  Organism  \\\n",
       "5       GCA_900215245.1_IMG-taxon_2617270901_annotated...  Bacteria   \n",
       "6           GCA_013388375.1_ASM1338837v1_cds_proteins.faa  Bacteria   \n",
       "7             GCA_000222975.1_ASM22297v1_cds_proteins.faa  Bacteria   \n",
       "31          GCA_022175585.1_ASM2217558v1_cds_proteins.faa  Bacteria   \n",
       "32           GCF_005156105.1_ASM515610v1_cds_proteins.faa  Bacteria   \n",
       "...                                                   ...       ...   \n",
       "114499      GCA_031236885.1_ASM3123688v1_cds_proteins.faa  Bacteria   \n",
       "114500      GCA_020673295.1_ASM2067329v1_cds_proteins.faa  Bacteria   \n",
       "114501      GCA_022488325.1_ASM2248832v1_cds_proteins.faa  Bacteria   \n",
       "114502      GCA_022488265.1_ASM2248826v1_cds_proteins.faa  Bacteria   \n",
       "114503      GCA_022488345.1_ASM2248834v1_cds_proteins.faa  Bacteria   \n",
       "\n",
       "                           Lineage  LineageTaxIDs          Strain    TaxID  \\\n",
       "5          Pseudomonas fluorescens            294      ATCC 13525      294   \n",
       "6           Xanthomonas campestris            339      MAFF106181   359385   \n",
       "7                  Yersinia pestis            632           A1122  1035377   \n",
       "31             Mycobacterium avium           1764  OCU889s_P11_4s   439334   \n",
       "32      Mycobacterium tuberculosis           1773     Danish 1331     1765   \n",
       "...                            ...            ...             ...      ...   \n",
       "114499            Escherichia coli            562            8523      562   \n",
       "114500            Escherichia coli            562            NDM7      562   \n",
       "114501            Escherichia coli            562    S-P-N-065.01      562   \n",
       "114502            Escherichia coli            562    S-P-N-045.01      562   \n",
       "114503            Escherichia coli            562    S-P-N-063.01      562   \n",
       "\n",
       "                      ComplexType Cluster  ...  NuoE  NuoF  NuoG  NuoH  NuoI  \\\n",
       "5       13 Subunits with Fused CD       3  ...  True  True  True  True  True   \n",
       "6            Complete 14 Subunits       3  ...  True  True  True  True  True   \n",
       "7       13 Subunits with Fused CD       6  ...  True  True  True  True  True   \n",
       "31           Complete 14 Subunits       2  ...  True  True  True  True  True   \n",
       "32           Complete 14 Subunits       6  ...  True  True  True  True  True   \n",
       "...                           ...     ...  ...   ...   ...   ...   ...   ...   \n",
       "114499  13 Subunits with Fused CD       4  ...  True  True  True  True  True   \n",
       "114500  13 Subunits with Fused CD       6  ...  True  True  True  True  True   \n",
       "114501  13 Subunits with Fused CD       8  ...  True  True  True  True  True   \n",
       "114502  13 Subunits with Fused CD       7  ...  True  True  True  True  True   \n",
       "114503  13 Subunits with Fused CD       8  ...  True  True  True  True  True   \n",
       "\n",
       "        NuoJ  NuoK  NuoL  NuoM  NuoN  \n",
       "5       True  True  True  True  True  \n",
       "6       True  True  True  True  True  \n",
       "7       True  True  True  True  True  \n",
       "31      True  True  True  True  True  \n",
       "32      True  True  True  True  True  \n",
       "...      ...   ...   ...   ...   ...  \n",
       "114499  True  True  True  True  True  \n",
       "114500  True  True  True  True  True  \n",
       "114501  True  True  True  True  True  \n",
       "114502  True  True  True  True  True  \n",
       "114503  True  True  True  True  True  \n",
       "\n",
       "[19961 rows x 26 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nuo_bool_detailed[(nuo_bool_detailed['ComplexType'].isin(['Complete 14 Subunits', '13 Subunits with Fused CD']))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "partial_comp_one_accession = nuo_bool_detailed[nuo_bool_detailed['ComplexType']=='Less than 11 Subunits']['Accession'].unique()\n",
    "partial_comp_one_hmmer = hmmer[hmmer['Accession'].isin(partial_comp_one_accession)]\n",
    "\n",
    "complete_comp_one_accession = nuo_bool_detailed[(nuo_bool_detailed['ComplexType'].isin(['Complete 14 Subunits', '13 Subunits with Fused CD']))]['Accession'].unique()\n",
    "complete_hmmer = hmmer[hmmer['Accession'].isin(complete_comp_one_accession)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Replicon</th>\n",
       "      <th>ProteomeFile</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Lineage</th>\n",
       "      <th>LineageTaxIDs</th>\n",
       "      <th>Strain</th>\n",
       "      <th>TaxID</th>\n",
       "      <th>ComplexType</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>...</th>\n",
       "      <th>NuoE</th>\n",
       "      <th>NuoF</th>\n",
       "      <th>NuoG</th>\n",
       "      <th>NuoH</th>\n",
       "      <th>NuoI</th>\n",
       "      <th>NuoJ</th>\n",
       "      <th>NuoK</th>\n",
       "      <th>NuoL</th>\n",
       "      <th>NuoM</th>\n",
       "      <th>NuoN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>U00096.3</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GCA_000005845.2_ASM584v2_cds_proteins.faa</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>562</td>\n",
       "      <td>K-12 substr. MG1655</td>\n",
       "      <td>511145</td>\n",
       "      <td>13 Subunits with Fused CD</td>\n",
       "      <td>8</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Accession    Replicon                               ProteomeFile  Organism  \\\n",
       "48  U00096.3  Chromosome  GCA_000005845.2_ASM584v2_cds_proteins.faa  Bacteria   \n",
       "\n",
       "             Lineage  LineageTaxIDs               Strain   TaxID  \\\n",
       "48  Escherichia coli            562  K-12 substr. MG1655  511145   \n",
       "\n",
       "                  ComplexType Cluster  ...  NuoE  NuoF  NuoG  NuoH  NuoI  \\\n",
       "48  13 Subunits with Fused CD       8  ...  True  True  True  True  True   \n",
       "\n",
       "    NuoJ  NuoK  NuoL  NuoM  NuoN  \n",
       "48  True  True  True  True  True  \n",
       "\n",
       "[1 rows x 26 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nuo_bool_detailed[nuo_bool_detailed['Accession']=='U00096.3']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter the nuo_bool_detailed DataFrame to include only the relevant ComplexTypes\n",
    "nuo_filtered = nuo_bool_detailed[nuo_bool_detailed['ComplexType'].isin(['Complete 14 Subunits', '13 Subunits with Fused CD'])]\n",
    "\n",
    "# Merge the filtered nuo_bool_detailed DataFrame with the complete_hmmer DataFrame\n",
    "complete_hmmer_filtered = complete_hmmer.merge(nuo_filtered[['Accession', 'Cluster']].drop_duplicates(), on=['Accession', 'Cluster'], how='inner')\n",
    "\n",
    "# Reset index of the resulting DataFrame\n",
    "complete_hmmer_filtered = complete_hmmer_filtered.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "41277"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hmmer['Accession'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19918"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "complete_hmmer_filtered['Accession'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17206"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "partial_comp_one_hmmer['Accession'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# saving_dir = \"/home/arglab/Work/Projects/04-Complex-I-Spread/Scripts-Data/MAIN-ANALYSIS/Secondary\"\n",
    "\n",
    "# nuo_count.to_csv(os.path.join(saving_dir, 'complex_one_count.csv'), index=False)\n",
    "\n",
    "# nuo_bool_detailed.to_csv(os.path.join(saving_dir, 'complex_one_boolean.csv'), index=False)\n",
    "\n",
    "# partial_comp_one_hmmer.to_csv(\"/home/arglab/Work/Projects/04-Complex-I-Spread/Scripts-Data/MAIN-ANALYSIS/processed_hmmer_search_results_02_partial.csv\", index=False)\n",
    "\n",
    "# complete_hmmer_filtered.to_csv(\"/home/arglab/Work/Projects/04-Complex-I-Spread/Scripts-Data/MAIN-ANALYSIS/processed_hmmer_search_results_02_complete.csv\", index=False)\n",
    "\n",
    "# hmmer.to_csv(\"/home/arglab/Work/Projects/04-Complex-I-Spread/Scripts-Data/MAIN-ANALYSIS/processed_hmmer_search_results_02_clustered.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def split_dataframe(df, n):\n",
    "#     # Calculate the number of subsets\n",
    "#     num_subsets = len(df) // n + (1 if len(df) % n != 0 else 0)\n",
    "#     subsets = []\n",
    "#     for i in range(num_subsets):\n",
    "#         subset = df.iloc[i*n:(i+1)*n]\n",
    "#         subsets.append(subset)\n",
    "#     return subsets\n",
    "\n",
    "# partial_comp_one_hmmer = partial_comp_one_hmmer.sort_values('ProteinAccession')\n",
    "\n",
    "# hmmer_subsets = split_dataframe(partial_comp_one_hmmer, 55000)\n",
    "\n",
    "# for i, subset in enumerate(hmmer_subsets):\n",
    "#     sequence_records = []\n",
    "#     subset = subset.groupby('ProteomeFile')\n",
    "#     for ProteomeFile, data in tqdm(subset, desc=f\"Fasta: {i+1}\"):\n",
    "#         ProteomeFile_path = os.path.join('/home/arglab/Work/Sequence-Data/ProteomeFiles', ProteomeFile)\n",
    "#         for record in SeqIO.parse(ProteomeFile_path, 'fasta'):\n",
    "#             if record.id in data['ProteinAccession'].unique().tolist():\n",
    "#                 sequence_record = SeqRecord(seq=Seq(str(record.seq).replace('*', '')), id=record.id, description=record.description)\n",
    "#                 sequence_records.append(sequence_record)\n",
    "\n",
    "#     SeqIO.write(sequence_records, os.path.join('/home/arglab/Work/Sequence-Data/', f\"potential_nuo_sequences_0{i}.faa\"), \"fasta\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
